# R script for CeNGEN analysis 11/23/20

# Uses as input the raw gene by barcode matrix generated by 3' UTR extension analysis.

BiocManager::install('multtest')
library(multtest)
devtools::install_github("constantAmateur/SoupX")
install.packages("prettyunits")

library(DropletUtils)
library(expss)
library(monocle)
library(Matrix)
library(scater)
library(Seurat)
library(SoupX)
library(ggplot2)

setwd("~/Dropbox (VU Basic Sciences)/Miller lab/10X Genomics")

# SoupX requires the raw_feature_bc_matrix and a filtered_feature_bc_matrix.
# Creating a new filtered feature by barcode matrix using emptydrops function from DropletUtils

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

my_dir <- "./final_gene_model/3697-ST-1/SoupX/raw_feature_bc_matrix"
unc_86 <- read10xCounts(my_dir)
colnames(unc_86) <- unc_86$Barcode
rowData(unc_86) <- rowData(unc_86)[,1:2]
gene_list <- read.table("./final_gene_model/3697-ST-1/SoupX/raw_feature_bc_matrix/features.tsv",
                        sep = "\t", stringsAsFactors = F, header = F)

rowData(unc_86)$Symbol <- as.character(vlookup(rowData(unc_86)$ID, gene_list))

bcrank <- barcodeRanks(counts(unc_86))
uniq <- !duplicated(bcrank$rank)
plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)

# Using droplets with fewer than 50 UMIs as the background. 
e.out <- emptyDrops(counts(unc_86), lower = 50)
summary(e.out$FDR <= 0.01)

# Mode   FALSE    TRUE    NA's 
# logical    7426    3896 6783558

# Checking what would be kept with default algorithm
c.keep <- defaultDrops(counts(unc_86), expected=5000)
e.keep <- e.out$FDR <=0.01

# Keeping cells from both default algorithm and emptydrops
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both   EmptyDrops
## 6791309        3571

#    Both EmptyDrops 
# 6791341       3539 

unc_86$Detection <- detection
unc_86$PValue <- e.out$PValue

unc_86 <- unc_86[, keep]
unc_86

table(unc_86$Detection)

unc_86_filtered_counts <- counts(unc_86)

# Generating new filtered matrix files for SoupX
system("gzip ./final_gene_model/3697-ST-1/SoupX/filtered_feature_bc_matrix/features.tsv")
writeMM(unc_86_filtered_counts, "./final_gene_model/3697-ST-1/SoupX2/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3697-ST-1/SoupX2/filtered_feature_bc_matrix/matrix.mtx")
writeMM(unc_86_filtered_counts, "./final_gene_model/3697-ST-1/SoupX2/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(unc_86)
write.table(fil.barcodes, "./final_gene_model/3697-ST-1/SoupX2/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3697-ST-1/SoupX2/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3697-ST-1/SoupX2/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(unc_86) <- librarySizeFactors(unc_86)
unc_86 <- normalize(unc_86)

unc86S <- as.Seurat(unc_86)

# Generating initial clustering with Seurat to aid in SoupX background correction
unc86S <- NormalizeData(unc86S)
unc86S <- FindVariableFeatures(unc86S, selection.method = "vst", nfeatures = 4000)
unc86S <- ScaleData(unc86S)
unc86S <- RunPCA(unc86S, features = VariableFeatures(unc86S), npcs = 100)

DimPlot(unc86S, reduction = "pca")
DimHeatmap(unc86S, dims = 1:15, cells = 500, balanced = TRUE)

ElbowPlot(unc86S, ndims = 100)
unc86S <- FindNeighbors(unc86S, dims = 1:100)
unc86S <- FindClusters(unc86S, resolution = 1.2)

unc86S <- RunUMAP(unc86S, dims = 1:100)
DimPlot(unc86S, reduction = "umap")
FeaturePlot(unc86S, features = "WBGene00003178")

class(unc86S@reductions$umap)
umap <- unc86S@reductions$umap

Embeddings(umap)[1:3,1:2]
unc86_DR <- as.data.frame(Embeddings(umap))
unc86_DR$Cluster <- Idents(unc86S)
head(unc86_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3697-ST-1/SoupX"
scl86 <- load10X(my_dir_S, soupRange = c(0, 25))

scl86$toc[1:5,1:5]
rownames(unc86_DR) <- substr(rownames(unc86_DR), 1, 16)
head(unc86_DR)

head(scl86$soupProfile[order(scl86$soupProfile$est, decreasing = TRUE), ], n = 20)

rownames(scl86$soupProfile) <- rownames(scl86$toc)
rownames(unc86_DR) <- paste(rownames(unc86_DR), "1", sep = "-")

gg = plotMarkerMap(scl86, "mec-17", unc86_DR)
plot(gg)

plotMarkerMap(scl86, geneSet = c("nlp-18", "mec-17"), unc86_DR)

scl86 <- setDR(scl86, unc86_DR)

gg <- plotMarkerMap(scl86, "mec-7")
plot(gg)

TRN.genes <- c("mec-17", "mec-7", "mec-3", "mec-10", "mec-12", "nlp-18")

useToEst = estimateNonExpressingCells(scl86, nonExpressedGeneList = list(TRN = TRN.genes), 
                                      clusters = setNames(unc86_DR$Cluster, rownames(unc86_DR)))
plotMarkerMap(scl86, geneSet = TRN.genes, DR = unc86_DR, useToEst = useToEst)
scl86 <- setClusters(scl86, unc86_DR$Cluster)

scl86 <- calculateContaminationFraction(scl86, list(TRN = TRN.genes), useToEst = useToEst)
# Estimated global contamination fraction of 6.45%

system.time(out <- adjustCounts(scl86, roundToInt = TRUE))

cntSoggy = rowSums(scl86$toc > 0)
cntStrained = rowSums(out > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(scl86$toc > out)/rowSums(scl86$toc > 0)), n = 20)

plotChangeMap(scl86, out, "mec-17")
plotChangeMap(scl86, out, "mec-7")
plotChangeMap(scl86, out, "col-124")
plotChangeMap(scl86, out, "col-140")
plotChangeMap(scl86, out, "snet-1")
plotChangeMap(scl86, out, "nlp-18")
plotChangeMap(scl86, out, "nduo-6")
plotChangeMap(scl86, out, "mec-12")

out[1:5,1:5]
scl86$toc[1:5,1:5]

saveRDS(scl86, "./final_gene_model/3697-ST-1/SoupX/soupX_object_unc86_1120.rds")
saveRDS(out, "./final_gene_model/3697-ST-1/SoupX/soupX_corrected_matrix_unc86_1120.rds")

dim(out)
head(colnames(out))
head(rownames(out))

# I need to make the rownames consistent also
rownames(out) <- as.character(vlookup(rownames(out), gene_list, result_column = 1, lookup_column = 2))
out[1:5,1:5]
head(colData(unc_86))

unc_86_corrected <- SingleCellExperiment(assays = list(counts = out), colData = colData(unc_86))  

rowData(unc_86_corrected) <- rowData(unc_86)

# Using scater to calculate some quality control metrics

unc_86 <- calculateQCMetrics(unc_86, feature_controls = list(Mito = which(rowData(unc_86)$ID == "WBGene00010959" | rowData(unc_86)$ID == "WBGene00010961"| rowData(unc_86)$ID == "WBGene00010966" | rowData(unc_86)$ID == "WBGene00010963" | rowData(unc_86)$ID =="WBGene00010957" | rowData(unc_86)$ID == "WBGene00010967" | rowData(unc_86)$ID == "WBGene00010958" | rowData(unc_86)$ID == "WBGene00010960" | rowData(unc_86)$ID == "WBGene00010962" | rowData(unc_86)$ID == "WBGene00000829" | rowData(unc_86)$ID =="WBGene00010965" | rowData(unc_86)$ID == "WBGene00010964")))
unc_86_corrected <- calculateQCMetrics(unc_86_corrected, feature_controls = list(Mito = which(rowData(unc_86_corrected)$ID == "WBGene00010959" | rowData(unc_86_corrected)$ID == "WBGene00010961"| rowData(unc_86_corrected)$ID == "WBGene00010966" | rowData(unc_86_corrected)$ID == "WBGene00010963" | rowData(unc_86_corrected)$ID =="WBGene00010957" | rowData(unc_86_corrected)$ID == "WBGene00010967" | rowData(unc_86_corrected)$ID == "WBGene00010958" | rowData(unc_86_corrected)$ID == "WBGene00010960" | rowData(unc_86_corrected)$ID == "WBGene00010962" | rowData(unc_86_corrected)$ID == "WBGene00000829" | rowData(unc_86_corrected)$ID =="WBGene00010965" | rowData(unc_86_corrected)$ID == "WBGene00010964")))

summary(unc_86$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.6164  1.1876  3.6981  2.7515 93.9492 

summary(unc_86_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.00000  0.07883  0.53648  3.07582  1.65589 94.21287 

table(unc_86$pct_counts_Mito < 20)
## FALSE   TRUE
##   113  3783 

table(unc_86_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  114  3782

# Visualizing general count distribution
par(mfrow=c(1,3))
hist(unc_86$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_86$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_86$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(unc_86_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_86_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_86_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(unc_86) <- librarySizeFactors(unc_86)
unc_86 <- normalize(unc_86)
unc_86
saveRDS(unc_86, "./final_gene_model/3697-ST-1/SoupX/020620_unc86_sce_uncorrected.rds")

sizeFactors(unc_86_corrected) <- librarySizeFactors(unc_86_corrected)
unc_86_corrected <- normalize(unc_86_corrected)
unc_86_corrected
saveRDS(unc_86_corrected, "./final_gene_model/3697-ST-1/SoupX/020620_unc86_sce_new_SoupX_corrected.rds")

# Next experiment

my_dir <- "./final_gene_model/1806-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

system("gzip ./final_gene_model/1806-ST-1/SoupX/raw_feature_bc_matrix/features.tsv")

pan_1 <- read10xCounts(my_dir)
colnames(pan_1) <- pan_1$Barcode
pan_1
head(rowData(pan_1))
rowData(pan_1) <- rowData(pan_1)[,1:2]

rowData(pan_1)$Symbol <- as.character(vlookup(rowData(pan_1)$ID, gene_list))

bcrank <- barcodeRanks(counts(pan_1))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(pan_1), lower = 50)
summary(e.out$FDR<=0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  8783    6938  721559 

c.keep <- defaultDrops(counts(pan_1), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
## 733371         24       3885 

pan_1$Detection <- detection
pan_1$PValue <- e.out$PValue

pan_1 <- pan_1[, keep]
pan_1

## class: SingleCellExperiment 
## dim: 46911 6962
## metadata(0):
##   assays(1): counts
## rownames(46911): WBGene00014450 WBGene00014451 ... WBGene00197051
## WBGene00199940
## rowData names(3): ID Symbol
## colnames(3896): AAACCCAAGAAGCCAC-1 AAACCCAAGCATGAAT-1 ...
## TTTGTTGAGCCTAGGA-1 TTTGTTGCAAGACGGT-1
## colData names(4): Sample Barcode Detection PValue
## reducedDimNames(0):
## spikeNames(0):

# Creating a filtered matrix to use for SoupX corrections

pan_1_filtered_counts <- counts(pan_1)
writeMM(pan_1_filtered_counts, "./final_gene_model/1806-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/1806-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(pan_1_filtered_counts, "./final_gene_model/1806-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(pan_1)
write.table(fil.barcodes, "./final_gene_model/1806-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/1806-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/1806-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(pan_1) <- librarySizeFactors(pan_1)
pan_1 <- normalize(pan_1)
pan_1S <- as.Seurat(pan_1)

pan_1S <- NormalizeData(pan_1S)
pan_1S <- FindVariableFeatures(pan_1S, selection.method = "vst", nfeatures = 4000)
pan_1S <- ScaleData(pan_1S)
pan_1S <- RunPCA(pan_1S, features = VariableFeatures(pan_1S), npcs = 130)

ElbowPlot(pan_1S, ndims = 130)
pan_1S <- FindNeighbors(pan_1S, dims = 1:130)
pan_1S <- FindClusters(pan_1S, resolution = 1.2)

pan_1S <- RunUMAP(pan_1S, dims = 1:130)
DimPlot(pan_1S, reduction = "umap")
FeaturePlot(pan_1S, features = "WBGene00003178")
FeaturePlot(pan_1S, features = "WBGene00001444")

umap.p <- pan_1S@reductions$umap

pan_1_DR <- as.data.frame(Embeddings(umap.p))
head(Idents(pan_1S))
pan_1_DR$Cluster <- Idents(pan_1S)
head(pan_1_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/1806-ST-1/SoupX"
sclpan1 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclpan1$toc[1:5,1:5]
head(rownames(pan_1_DR))

rownames(pan_1_DR) <- substr(rownames(pan_1_DR), 1, 16)
head(pan_1_DR)
pan_1_DR$mec17 <- sclpan1$toc["WBGene00001444", rownames(pan_1_DR)]
gg <- ggplot(pan_1_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = mec17 > 0))
plot(gg)

head(sclpan1$soupProfile[order(sclpan1$soupProfile$est, decreasing = TRUE), ], n = 20)

rownames(sclpan1$toc) <- as.character(vlookup(rownames(sclpan1$toc), gene_list))
sclpan1$toc[1:5,1:5]
rownames(sclpan1$soupProfile) <- rownames(sclpan1$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclpan1, "flp-1", pan_1_DR)
plot(gg)

plotMarkerMap(sclpan1, geneSet = c("flp-1", "nlp-18", "snet-1", "nlp-17"), pan_1_DR)
head(sclpan1$soupProfile[order(sclpan1$soupProfile$est, decreasing = TRUE), ], n = 20)

# Now it works.

head(pan_1_DR)
sclpan1 <- setDR(sclpan1, pan_1_DR)

gg <- plotMarkerMap(sclpan1, "nlp-49")
plot(gg)

sclpan1 <- setClusters(sclpan1, pan_1_DR$Cluster)
plotMarkerDistribution(sclpan1)
pan.1.genes <- c("flp-1", "snet-1", "nlp-49")

useToEst = estimateNonExpressingCells(sclpan1, nonExpressedGeneList = list(pan1 = pan.1.genes))
plotMarkerMap(sclpan1, geneSet = pan.1.genes, DR = pan_1_DR, useToEst = useToEst)

sclpan1 <- calculateContaminationFraction(sclpan1, list(pan1 = pan.1.genes), useToEst = useToEst)
# Estimated global contamination fraction of 6.89%

system.time(out.p <- adjustCounts(sclpan1, roundToInt = TRUE))

cntSoggy = rowSums(sclpan1$toc > 0)
cntStrained = rowSums(out > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclpan1$toc > out.p)/rowSums(sclpan1$toc > 0)), n = 20)

plotChangeMap(sclpan1, out.p, "mec-17")
plotChangeMap(sclpan1, out.p, "flp-1")
plotChangeMap(sclpan1, out.p, "snet-1")
plotChangeMap(sclpan1, out.p, "flp-9")
plotChangeMap(sclpan1, out.p, "nlp-49")
plotChangeMap(sclpan1, out.p, "nlp-17")
plotChangeMap(sclpan1, out.p, "nduo-6")
plotChangeMap(sclpan1, out.p, "mec-12")
plotChangeMap(sclpan1, out.p, "sre-46")
plotChangeMap(sclpan1, out.p, "sre-43")

out.p[1:5,1:5]
sclpan1$toc[1:5,1:5]

saveRDS(sclpan1, "./final_gene_model/1806-ST-1/SoupX/soupX_object_pan_1.rds")
saveRDS(out.p, "./final_gene_model/1806-ST-1/SoupX/soupX_corrected_matrix_pan_1.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data
dim(out.p)
head(fil.barcodes)
colnames(out.p) <- paste(colnames(out.p), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.p) <- as.character(vlookup(rownames(out.p), gene_list, result_column = 1, lookup_column = 2))
out.p[1:5,1:5]
head(colData(pan_1))

pan_1_corrected <- SingleCellExperiment(assays = list(counts = out.p), colData = colData(pan_1))  
pan_1
pan_1_corrected
head(rowData(pan_1))
head(rowData(pan_1_corrected))
rowData(pan_1_corrected) <- rowData(pan_1)
pan_1_corrected

pan_1 <- calculateQCMetrics(pan_1, feature_controls = list(Mito = which(rowData(pan_1)$ID == "WBGene00010959" | rowData(pan_1)$ID == "WBGene00010961"| rowData(pan_1)$ID == "WBGene00010966" | rowData(pan_1)$ID == "WBGene00010963" | rowData(pan_1)$ID =="WBGene00010957" | rowData(pan_1)$ID == "WBGene00010967" | rowData(pan_1)$ID == "WBGene00010958" | rowData(pan_1)$ID == "WBGene00010960" | rowData(pan_1)$ID == "WBGene00010962" | rowData(pan_1)$ID == "WBGene00000829" | rowData(pan_1)$ID =="WBGene00010965" | rowData(pan_1)$ID == "WBGene00010964")))
pan_1_corrected <- calculateQCMetrics(pan_1_corrected, feature_controls = list(Mito = which(rowData(pan_1_corrected)$ID == "WBGene00010959" | rowData(pan_1_corrected)$ID == "WBGene00010961"| rowData(pan_1_corrected)$ID == "WBGene00010966" | rowData(pan_1_corrected)$ID == "WBGene00010963" | rowData(pan_1_corrected)$ID =="WBGene00010957" | rowData(pan_1_corrected)$ID == "WBGene00010967" | rowData(pan_1_corrected)$ID == "WBGene00010958" | rowData(pan_1_corrected)$ID == "WBGene00010960" | rowData(pan_1_corrected)$ID == "WBGene00010962" | rowData(pan_1_corrected)$ID == "WBGene00000829" | rowData(pan_1_corrected)$ID =="WBGene00010965" | rowData(pan_1_corrected)$ID == "WBGene00010964")))

summary(pan_1$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   2.823   6.167  11.118  15.974  84.211

summary(pan_1_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## .0000  0.6695  2.9040  8.0428 10.9091 85.2321 

table(pan_1$pct_counts_Mito < 20)
## FALSE   TRUE
##   1345  5617

table(pan_1_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  975  5987

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(pan_1$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(pan_1$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(pan_1$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(pan_1_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(pan_1_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(pan_1_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(pan_1) <- librarySizeFactors(pan_1)
pan_1 <- normalize(pan_1)
pan_1
saveRDS(pan_1, "./final_gene_model/1806-ST-1/SoupX/020620_pan_1_sce_uncorrected.rds")

sizeFactors(pan_1_corrected) <- librarySizeFactors(pan_1_corrected)
pan_1_corrected <- normalize(pan_1_corrected)
pan_1_corrected
saveRDS(pan_1_corrected, "./final_gene_model/1806-ST-1/SoupX/020620_pan_1_sce_new_SoupX_corrected.rds")

# 1806-2

my_dir <- "./final_gene_model/1806-ST-2/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

pan_2 <- read10xCounts(my_dir)
colnames(pan_2) <- pan_2$Barcode
pan_2
head(rowData(pan_2))
rowData(pan_2) <- rowData(pan_2)[,1:2]

rowData(pan_2)$Symbol <- as.character(vlookup(rowData(pan_2)$ID, gene_list))
head(rowData(pan_2))

bcrank <- barcodeRanks(counts(pan_2))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(pan_2), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical 10424    9361  717495

c.keep <- defaultDrops(counts(pan_2), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
## 732263         25       4992

pan_2$Detection <- detection
pan_2$PValue <- e.out$PValue

pan_2 <- pan_2[, keep]
pan_2
## class: SingleCellExperiment 
## dim: 46911 9386
## metadata(0):
##   assays(1): counts
## rownames(46911): WBGene00014450 WBGene00014451 ... WBGene00197051
## WBGene00199940
## rowData names(3): ID Symbol
## colnames(9386): AAACCCAAGAAGCCAC-1 AAACCCAAGCATGAAT-1 ...
## TTTGTTGAGCCTAGGA-1 TTTGTTGCAAGACGGT-1
## colData names(4): Sample Barcode Detection PValue
## reducedDimNames(0):
## spikeNames(0):

# Creating a filtered matrix to use for SoupX corrections

pan_2_filtered_counts <- counts(pan_2)
writeMM(pan_2_filtered_counts, "./final_gene_model/1806-ST-2/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/1806-ST-2/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(pan_2_filtered_counts, "./final_gene_model/1806-ST-2/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(pan_2)
write.table(fil.barcodes, "./final_gene_model/1806-ST-2/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/1806-ST-2/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/1806-ST-2/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(pan_2) <- librarySizeFactors(pan_2)
pan_2 <- normalize(pan_2)
pan_2S <- as.Seurat(pan_2)

pan_2S <- NormalizeData(pan_2S)
pan_2S <- FindVariableFeatures(pan_2S, selection.method = "vst", nfeatures = 4000)
pan_2S <- ScaleData(pan_2S)
pan_2S <- RunPCA(pan_2S, features = VariableFeatures(pan_2S), npcs = 130)

ElbowPlot(pan_2S, ndims = 130)

pan_2S <- FindNeighbors(pan_2S, dims = 1:130)
pan_2S <- FindClusters(pan_2S, resolution = 1.2)

pan_2S <- RunUMAP(pan_2S, dims = 1:130)
DimPlot(pan_2S, reduction = "umap")
FeaturePlot(pan_2S, features = "WBGene00003178")
FeaturePlot(pan_2S, features = "WBGene00001444")

umap.p2 <- pan_2S@reductions$umap

pan_2_DR <- as.data.frame(Embeddings(umap.p2))
head(Idents(pan_2S))
pan_2_DR$Cluster <- Idents(pan_2S)
head(pan_2_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/1806-ST-2/SoupX"
sclpan2 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclpan2$toc[1:5,1:5]
head(rownames(pan_2_DR))

rownames(pan_2_DR) <- substr(rownames(pan_2_DR), 1, 16)
head(pan_2_DR)
pan_2_DR$flp1 <- sclpan2$toc["WBGene00001444", rownames(pan_2_DR)]
gg <- ggplot(pan_2_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = flp1 > 0))
plot(gg)

rownames(sclpan2$toc) <- as.character(vlookup(rownames(sclpan2$toc), gene_list))
sclpan2$toc[1:5,1:5]
rownames(sclpan2$soupProfile) <- rownames(sclpan2$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclpan2, "flp-1", pan_2_DR)
plot(gg)

plotMarkerMap(sclpan2, geneSet = c("flp-1", "nlp-18", "snet-1", "nlp-17"), pan_2_DR)
head(sclpan2$soupProfile[order(sclpan2$soupProfile$est, decreasing = TRUE), ], n = 20)

# Now it works.

head(pan_2_DR)
sclpan2 <- setDR(sclpan2, pan_2_DR)

gg <- plotMarkerMap(sclpan2, "nlp-49")
plot(gg)

sclpan2 <- setClusters(sclpan2, pan_2_DR$Cluster)
plotMarkerDistribution(sclpan2)
pan2.genes <- c("flp-1", "snet-1", "nlp-49")

useToEst = estimateNonExpressingCells(sclpan2, nonExpressedGeneList = list(pan2 = pan2.genes))
plotMarkerMap(sclpan2, geneSet = pan2.genes, DR = pan_2_DR, useToEst = useToEst)

sclpan2 <- calculateContaminationFraction(sclpan2, list(pan2 = pan2.genes), useToEst = useToEst)
# Estimated global contamination fraction of 10.06%

system.time(out.p2 <- adjustCounts(sclpan2, roundToInt = TRUE))
system.time(out.p2.ni <- adjustCounts(sclpan2))

# I will also generate now the non-rounded outputs of scl86 and sclpan1, to have both
system.time(out.ni <- adjustCounts(scl86))
system.time(out.p.ni <- adjustCounts(sclpan1))


cntSoggy = rowSums(sclpan2$toc > 0)
cntStrained = rowSums(out > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclpan2$toc > out.p2)/rowSums(sclpan2$toc > 0)), n = 20)

plotChangeMap(sclpan2, out.p2, "mec-17")
plotChangeMap(sclpan2, out.p2, "flp-1")
plotChangeMap(sclpan2, out.p2, "snet-1")
plotChangeMap(sclpan2, out.p2, "flp-9")
plotChangeMap(sclpan2, out.p2, "nlp-49")
plotChangeMap(sclpan2, out.p2, "nlp-17")
plotChangeMap(sclpan2, out.p2, "nduo-6")
plotChangeMap(sclpan2, out.p2, "mec-12")
plotChangeMap(sclpan2, out.p2, "sre-46")
plotChangeMap(sclpan2, out.p2, "sre-43")

out.p2[1:5,1:5]
sclpan2$toc[1:5,1:5]

saveRDS(sclpan2, "./final_gene_model/1806-ST-2/SoupX/soupX_object_pan_2.rds")
saveRDS(out.p2, "./final_gene_model/1806-ST-2/SoupX/soupX_corrected_matrix_pan_2.rds")
saveRDS(out.p2.ni, "./final_gene_model/1806-ST-2/SoupX/soupX_corrected_matrix_pan_2_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data
dim(out.p2)
head(fil.barcodes)
colnames(out.p2) <- paste(colnames(out.p2), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.p2) <- as.character(vlookup(rownames(out.p2), gene_list, result_column = 1, lookup_column = 2))
out.p2[1:5,1:5]
head(colData(pan_2))

pan_2_corrected <- SingleCellExperiment(assays = list(counts = out.p2), colData = colData(pan_2))  
pan_2
pan_2_corrected
head(rowData(pan_2))
head(rowData(pan_2_corrected))
rowData(pan_2_corrected) <- rowData(pan_2)
pan_2_corrected

pan_2 <- calculateQCMetrics(pan_2, feature_controls = list(Mito = which(rowData(pan_2)$ID == "WBGene00010959" | rowData(pan_2)$ID == "WBGene00010961"| rowData(pan_2)$ID == "WBGene00010966" | rowData(pan_2)$ID == "WBGene00010963" | rowData(pan_2)$ID =="WBGene00010957" | rowData(pan_2)$ID == "WBGene00010967" | rowData(pan_2)$ID == "WBGene00010958" | rowData(pan_2)$ID == "WBGene00010960" | rowData(pan_2)$ID == "WBGene00010962" | rowData(pan_2)$ID == "WBGene00000829" | rowData(pan_2)$ID =="WBGene00010965" | rowData(pan_2)$ID == "WBGene00010964")))
pan_2_corrected <- calculateQCMetrics(pan_2_corrected, feature_controls = list(Mito = which(rowData(pan_2_corrected)$ID == "WBGene00010959" | rowData(pan_2_corrected)$ID == "WBGene00010961"| rowData(pan_2_corrected)$ID == "WBGene00010966" | rowData(pan_2_corrected)$ID == "WBGene00010963" | rowData(pan_2_corrected)$ID =="WBGene00010957" | rowData(pan_2_corrected)$ID == "WBGene00010967" | rowData(pan_2_corrected)$ID == "WBGene00010958" | rowData(pan_2_corrected)$ID == "WBGene00010960" | rowData(pan_2_corrected)$ID == "WBGene00010962" | rowData(pan_2_corrected)$ID == "WBGene00000829" | rowData(pan_2_corrected)$ID =="WBGene00010965" | rowData(pan_2_corrected)$ID == "WBGene00010964")))

summary(pan_2$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   3.191   6.665  11.149  16.297  90.000 

summary(pan_2_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.4756  2.7211  7.5928 10.4392 91.1612 

table(pan_2$pct_counts_Mito < 20)
## FALSE   TRUE
##  1827  7559 

table(pan_2_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  1253  8133 

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(pan_2$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(pan_2$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(pan_2$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(pan_2_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(pan_2_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(pan_2_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(pan_2) <- librarySizeFactors(pan_2)
pan_2 <- normalize(pan_2)
pan_2
saveRDS(pan_2, "./final_gene_model/1806-ST-2/SoupX/020620_pan_2_sce_uncorrected.rds")

sizeFactors(pan_2_corrected) <- librarySizeFactors(pan_2_corrected)
pan_2_corrected <- normalize(pan_2_corrected)
pan_2_corrected
saveRDS(pan_2_corrected, "./final_gene_model/1806-ST-2/SoupX/020620_pan_2_sce_new_SoupX_corrected.rds")

saveRDS(out.ni, "./final_gene_model/3697-ST-1/SoupX/020620_unc86_corrected_matrix_non_integer.rds")
saveRDS(out.p.ni, "./final_gene_model/3697-ST-1/SoupX/020620_pan_1_corrected_matrix_non_integer.rds")


# 2658-ST-1

my_dir <- "./final_gene_model/2658-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

unc_47_1 <- read10xCounts(my_dir)
colnames(unc_47_1) <- unc_47_1$Barcode
unc_47_1
head(rowData(unc_47_1))
rowData(unc_47_1) <- rowData(unc_47_1)[,1:2]

rowData(unc_47_1)$Symbol <- as.character(vlookup(rowData(unc_47_1)$ID, gene_list))
head(rowData(unc_47_1))

bcrank <- barcodeRanks(counts(unc_47_1))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(unc_47_1), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical 1392    1680  734208 

c.keep <- defaultDrops(counts(unc_47_1), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
##  736718         97        465 

unc_47_1$Detection <- detection
unc_47_1$PValue <- e.out$PValue

unc_47_1 <- unc_47_1[, keep]
unc_47_1
## class: SingleCellExperiment 
## dim: 46911 1777
## metadata(0):
##   assays(1): counts
## rownames(46911): WBGene00014450 WBGene00014451 ... WBGene00197051
## WBGene00199940
## rowData names(3): ID Symbol
## colnames(1777): AAACCCAAGAAGCCAC-1 AAACCCAAGCATGAAT-1 ...
## TTTGTTGAGCCTAGGA-1 TTTGTTGCAAGACGGT-1
## colData names(4): Sample Barcode Detection PValue
## reducedDimNames(0):
## spikeNames(0):

# Creating a filtered matrix to use for SoupX corrections

unc_47_1_filtered_counts <- counts(unc_47_1)
writeMM(unc_47_1_filtered_counts, "./final_gene_model/2658-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/2658-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(unc_47_1_filtered_counts, "./final_gene_model/2658-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(unc_47_1)
write.table(fil.barcodes, "./final_gene_model/2658-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/2658-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/2658-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(unc_47_1) <- librarySizeFactors(unc_47_1)
unc_47_1 <- normalize(unc_47_1)
unc_47_1S <- as.Seurat(unc_47_1)

unc_47_1S <- NormalizeData(unc_47_1S)
unc_47_1S <- FindVariableFeatures(unc_47_1S, selection.method = "vst", nfeatures = 4000)
unc_47_1S <- ScaleData(unc_47_1S)
unc_47_1S <- RunPCA(unc_47_1S, features = VariableFeatures(unc_47_1S), npcs = 100)

ElbowPlot(unc_47_1S, ndims = 100)

unc_47_1S <- FindNeighbors(unc_47_1S, dims = 1:100)
unc_47_1S <- FindClusters(unc_47_1S, resolution = 1.2)

unc_47_1S <- RunUMAP(unc_47_1S, dims = 1:100)
DimPlot(unc_47_1S, reduction = "umap")
FeaturePlot(unc_47_1S, features = "WBGene00007304")

umap.47_1 <- unc_47_1S@reductions$umap

unc_47_1_DR <- as.data.frame(Embeddings(umap.47_1))
head(Idents(unc_47_1S))
unc_47_1_DR$Cluster <- Idents(unc_47_1S)
head(unc_47_1_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/2658-ST-1/SoupX"
sclunc47.1 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclunc47.1$toc[1:5,1:5]
head(rownames(unc_47_1_DR))

rownames(unc_47_1_DR) <- substr(rownames(unc_47_1_DR), 1, 16)
head(unc_47_1_DR)
unc_47_1_DR$ttr39 <- sclunc47.1$toc["WBGene00007304", rownames(unc_47_1_DR)]
gg <- ggplot(unc_47_1_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = ttr39 > 0))
plot(gg)

rownames(sclunc47.1$toc) <- as.character(vlookup(rownames(sclunc47.1$toc), gene_list))
sclunc47.1$toc[1:5,1:5]
rownames(sclunc47.1$soupProfile) <- rownames(sclunc47.1$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclunc47.1, "ttr-39", unc_47_1_DR)
plot(gg)
plotMarkerMap(sclunc47.1, "flp-9", unc_47_1_DR)
plotMarkerMap(sclunc47.1, "T05A8.3", unc_47_1_DR)


plotMarkerMap(sclunc47.1, geneSet = c("ttr-39","flp-11","flp-9"), unc_47_1_DR)
head(sclunc47.1$soupProfile[order(sclunc47.1$soupProfile$est, decreasing = TRUE), ], n = 20)

# Now it works.

head(unc_47_1_DR)
sclunc47.1 <- setDR(sclunc47.1, unc_47_1_DR)

sclunc47.1 <- setClusters(sclunc47.1, unc_47_1_DR$Cluster)
plotMarkerDistribution(sclunc47.1)
unc47.genes <- c("flp-11", "ttr-39", "flp-9", "T05A8.3")

useToEst = estimateNonExpressingCells(sclunc47.1, nonExpressedGeneList = list(unc47 = unc47.genes))
plotMarkerMap(sclunc47.1, geneSet = unc47.genes, DR = unc_47_1_DR, useToEst = useToEst)

sclunc47.1 <- calculateContaminationFraction(sclunc47.1, list(unc47 = unc47.genes), useToEst = useToEst)
# Estimated global contamination fraction of 10.79%

system.time(out.47.1 <- adjustCounts(sclunc47.1, roundToInt = TRUE))
system.time(out.47.1.ni <- adjustCounts(sclunc47.1))

cntSoggy = rowSums(sclunc47.1$toc > 0)
cntStrained = rowSums(out.47.1 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclunc47.1$toc > out.47.1)/rowSums(sclunc47.1$toc > 0)), n = 20)

plotChangeMap(sclunc47.1, out.47.1, "ttr-39")
plotChangeMap(sclunc47.1, out.47.1, "flp-9")
plotChangeMap(sclunc47.1, out.47.1, "flp-11")
plotChangeMap(sclunc47.1, out.47.1, "T05A8.3")
plotChangeMap(sclunc47.1, out.47.1, "unc-25")
plotChangeMap(sclunc47.1, out.47.1, "K09F6.13")
plotChangeMap(sclunc47.1, out.47.1, "nduo-6")

out.47.1[1:5,1:5]
sclunc47.1$toc[1:5,1:5]

saveRDS(sclunc47.1, "./final_gene_model/2658-ST-1/SoupX/soupX_object_unc_47_1.rds")
saveRDS(out.47.1, "./final_gene_model/2658-ST-1/SoupX/soupX_corrected_matrix_unc_47_1.rds")
saveRDS(out.47.1.ni, "./final_gene_model/2658-ST-1/SoupX/soupX_corrected_matrix_unc_47_1_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.47.1) <- paste(colnames(out.47.1), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.47.1) <- as.character(vlookup(rownames(out.47.1), gene_list, result_column = 1, lookup_column = 2))
out.47.1[1:5,1:5]
head(colData(unc_47_1))

unc_47_1_corrected <- SingleCellExperiment(assays = list(counts = out.47.1), colData = colData(unc_47_1))  
unc_47_1
unc_47_1_corrected
head(rowData(unc_47_1))
head(rowData(unc_47_1_corrected))
rowData(unc_47_1_corrected) <- rowData(unc_47_1)
unc_47_1_corrected

unc_47_1 <- calculateQCMetrics(unc_47_1, feature_controls = list(Mito = which(rowData(unc_47_1)$ID == "WBGene00010959" | rowData(unc_47_1)$ID == "WBGene00010961"| rowData(unc_47_1)$ID == "WBGene00010966" | rowData(unc_47_1)$ID == "WBGene00010963" | rowData(unc_47_1)$ID =="WBGene00010957" | rowData(unc_47_1)$ID == "WBGene00010967" | rowData(unc_47_1)$ID == "WBGene00010958" | rowData(unc_47_1)$ID == "WBGene00010960" | rowData(unc_47_1)$ID == "WBGene00010962" | rowData(unc_47_1)$ID == "WBGene00000829" | rowData(unc_47_1)$ID =="WBGene00010965" | rowData(unc_47_1)$ID == "WBGene00010964")))
unc_47_1_corrected <- calculateQCMetrics(unc_47_1_corrected, feature_controls = list(Mito = which(rowData(unc_47_1_corrected)$ID == "WBGene00010959" | rowData(unc_47_1_corrected)$ID == "WBGene00010961"| rowData(unc_47_1_corrected)$ID == "WBGene00010966" | rowData(unc_47_1_corrected)$ID == "WBGene00010963" | rowData(unc_47_1_corrected)$ID =="WBGene00010957" | rowData(unc_47_1_corrected)$ID == "WBGene00010967" | rowData(unc_47_1_corrected)$ID == "WBGene00010958" | rowData(unc_47_1_corrected)$ID == "WBGene00010960" | rowData(unc_47_1_corrected)$ID == "WBGene00010962" | rowData(unc_47_1_corrected)$ID == "WBGene00000829" | rowData(unc_47_1_corrected)$ID =="WBGene00010965" | rowData(unc_47_1_corrected)$ID == "WBGene00010964")))

summary(unc_47_1$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.887   3.763   7.134   8.280  96.721 

summary(unc_47_1_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   0.000   1.199   4.864   4.930  97.954

table(unc_47_1$pct_counts_Mito < 20)
## FALSE   TRUE
##  119  1658

table(unc_47_1_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##   86  1691

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(unc_47_1$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_47_1$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_47_1$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(unc_47_1_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_47_1_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_47_1_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(unc_47_1) <- librarySizeFactors(unc_47_1)
unc_47_1 <- normalize(unc_47_1)
unc_47_1
saveRDS(unc_47_1, "./final_gene_model/2658-ST-1/SoupX/020620_unc_47_1_sce_uncorrected.rds")

sizeFactors(unc_47_1_corrected) <- librarySizeFactors(unc_47_1_corrected)
unc_47_1_corrected <- normalize(unc_47_1_corrected)
unc_47_1_corrected
saveRDS(unc_47_1_corrected, "./final_gene_model/2658-ST-1/SoupX/020620_unc_47_1_sce_new_SoupX_corrected.rds")

# 3131-ST-1

my_dir <- "./final_gene_model/3131-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

unc_47_2 <- read10xCounts(my_dir)
colnames(unc_47_2) <- unc_47_2$Barcode
unc_47_2
head(rowData(unc_47_2))
rowData(unc_47_2) <- rowData(unc_47_2)[,1:2]

rowData(unc_47_2)$Symbol <- as.character(vlookup(rowData(unc_47_2)$ID, gene_list))
head(rowData(unc_47_2))

bcrank <- barcodeRanks(counts(unc_47_2))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(unc_47_2), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  3598    4021  729661  

c.keep <- defaultDrops(counts(unc_47_2), expected=5000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
##  735970         23       1287

unc_47_2$Detection <- detection
unc_47_2$PValue <- e.out$PValue

unc_47_2 <- unc_47_2[, keep]
unc_47_2
## class: SingleCellExperiment 
## dim: 46911 4044

# Creating a filtered matrix to use for SoupX corrections

unc_47_2_filtered_counts <- counts(unc_47_2)
writeMM(unc_47_2_filtered_counts, "./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(unc_47_2_filtered_counts, "./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(unc_47_2)
write.table(fil.barcodes, "./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(unc_47_2) <- librarySizeFactors(unc_47_2)
unc_47_2 <- normalize(unc_47_2)
unc_47_2S <- as.Seurat(unc_47_2)

unc_47_2S <- NormalizeData(unc_47_2S)
unc_47_2S <- FindVariableFeatures(unc_47_2S, selection.method = "vst", nfeatures = 4000)
unc_47_2S <- ScaleData(unc_47_2S)
unc_47_2S <- RunPCA(unc_47_2S, features = VariableFeatures(unc_47_2S), npcs = 125)

ElbowPlot(unc_47_2S, ndims = 125)

unc_47_2S <- FindNeighbors(unc_47_2S, dims = 1:120)
unc_47_2S <- FindClusters(unc_47_2S, resolution = 1.2)

unc_47_2S <- RunUMAP(unc_47_2S, dims = 1:120)
DimPlot(unc_47_2S, reduction = "umap")
FeaturePlot(unc_47_2S, features = "WBGene00007304")

umap.47_2 <- unc_47_2S@reductions$umap

unc_47_2_DR <- as.data.frame(Embeddings(umap.47_2))
head(Idents(unc_47_2S))
unc_47_2_DR$Cluster <- Idents(unc_47_2S)
head(unc_47_2_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3131-ST-1/SoupX"
sclunc47.2 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclunc47.2$toc[1:5,1:5]
head(rownames(unc_47_2_DR))

rownames(unc_47_2_DR) <- substr(rownames(unc_47_2_DR), 1, 16)
head(unc_47_2_DR)
unc_47_2_DR$ttr39 <- sclunc47.2$toc["WBGene00007304", rownames(unc_47_2_DR)]
gg <- ggplot(unc_47_2_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = ttr39 > 0))
plot(gg)

rownames(sclunc47.2$toc) <- as.character(vlookup(rownames(sclunc47.2$toc), gene_list))
sclunc47.2$toc[1:5,1:5]
rownames(sclunc47.2$soupProfile) <- rownames(sclunc47.2$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclunc47.2, "ttr-39", unc_47_2_DR)
plot(gg)
plotMarkerMap(sclunc47.2, "flp-9", unc_47_2_DR)
plotMarkerMap(sclunc47.2, "K09F6.13", unc_47_2_DR)


plotMarkerMap(sclunc47.2, geneSet = c("ttr-39","flp-11","flp-9"), unc_47_2_DR)
head(sclunc47.2$soupProfile[order(sclunc47.2$soupProfile$est, decreasing = TRUE), ], n = 20)

# Now it works.

head(unc_47_2_DR)
sclunc47.2 <- setDR(sclunc47.2, unc_47_2_DR)

sclunc47.2 <- setClusters(sclunc47.2, unc_47_2_DR$Cluster)
plotMarkerDistribution(sclunc47.2)
unc47.genes <- c("ttr-39", "flp-9", "flp-8", "flp-11")

useToEst = estimateNonExpressingCells(sclunc47.2, nonExpressedGeneList = list(unc47 = unc47.genes))
plotMarkerMap(sclunc47.2, geneSet = unc47.genes, DR = unc_47_2_DR, useToEst = useToEst)

sclunc47.2 <- calculateContaminationFraction(sclunc47.2, list(unc47 = unc47.genes), useToEst = useToEst)
# Estimated global contamination fraction of 13.56%

system.time(out.47.2 <- adjustCounts(sclunc47.2, roundToInt = TRUE))
system.time(out.47.2.ni <- adjustCounts(sclunc47.2))

cntSoggy = rowSums(sclunc47.2$toc > 0)
cntStrained = rowSums(out.47.2 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclunc47.2$toc > out.47.2)/rowSums(sclunc47.2$toc > 0)), n = 20)

plotChangeMap(sclunc47.2, out.47.2, "ttr-39")
plotChangeMap(sclunc47.2, out.47.2, "flp-9")
plotChangeMap(sclunc47.2, out.47.2, "flp-11")
plotChangeMap(sclunc47.2, out.47.2, "T05A8.3")
plotChangeMap(sclunc47.2, out.47.2, "unc-25")
plotChangeMap(sclunc47.2, out.47.2, "K09F6.13")
plotChangeMap(sclunc47.2, out.47.2, "nduo-6")

out.47.2[1:5,1:5]
sclunc47.2$toc[1:5,1:5]

saveRDS(sclunc47.2, "./final_gene_model/3131-ST-1/SoupX/soupX_object_unc_47_2.rds")
saveRDS(out.47.2, "./final_gene_model/3131-ST-1/SoupX/soupX_corrected_matrix_unc_47_2.rds")
saveRDS(out.47.2.ni, "./final_gene_model/3131-ST-1/SoupX/soupX_corrected_matrix_unc_47_2_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.47.2) <- paste(colnames(out.47.2), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.47.2) <- as.character(vlookup(rownames(out.47.2), gene_list, result_column = 1, lookup_column = 2))
out.47.2[1:5,1:5]
head(colData(unc_47_2))

unc_47_2_corrected <- SingleCellExperiment(assays = list(counts = out.47.2), colData = colData(unc_47_2))  
unc_47_2
unc_47_2_corrected
head(rowData(unc_47_2))
head(rowData(unc_47_2_corrected))
rowData(unc_47_2_corrected) <- rowData(unc_47_2)
unc_47_2_corrected

unc_47_2 <- calculateQCMetrics(unc_47_2, feature_controls = list(Mito = which(rowData(unc_47_2)$ID == "WBGene00010959" | rowData(unc_47_2)$ID == "WBGene00010961"| rowData(unc_47_2)$ID == "WBGene00010966" | rowData(unc_47_2)$ID == "WBGene00010963" | rowData(unc_47_2)$ID =="WBGene00010957" | rowData(unc_47_2)$ID == "WBGene00010967" | rowData(unc_47_2)$ID == "WBGene00010958" | rowData(unc_47_2)$ID == "WBGene00010960" | rowData(unc_47_2)$ID == "WBGene00010962" | rowData(unc_47_2)$ID == "WBGene00000829" | rowData(unc_47_2)$ID =="WBGene00010965" | rowData(unc_47_2)$ID == "WBGene00010964")))
unc_47_2_corrected <- calculateQCMetrics(unc_47_2_corrected, feature_controls = list(Mito = which(rowData(unc_47_2_corrected)$ID == "WBGene00010959" | rowData(unc_47_2_corrected)$ID == "WBGene00010961"| rowData(unc_47_2_corrected)$ID == "WBGene00010966" | rowData(unc_47_2_corrected)$ID == "WBGene00010963" | rowData(unc_47_2_corrected)$ID =="WBGene00010957" | rowData(unc_47_2_corrected)$ID == "WBGene00010967" | rowData(unc_47_2_corrected)$ID == "WBGene00010958" | rowData(unc_47_2_corrected)$ID == "WBGene00010960" | rowData(unc_47_2_corrected)$ID == "WBGene00010962" | rowData(unc_47_2_corrected)$ID == "WBGene00000829" | rowData(unc_47_2_corrected)$ID =="WBGene00010965" | rowData(unc_47_2_corrected)$ID == "WBGene00010964")))

summary(unc_47_2$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   2.014   3.601   6.315   7.224  98.850 

summary(unc_47_2_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.0000  0.4468  3.1004  2.3923 99.1474 

table(unc_47_2$pct_counts_Mito < 20)
## FALSE  TRUE
##  248  3796

table(unc_47_2_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  155  3889 

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(unc_47_2$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_47_2$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_47_2$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(unc_47_2_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_47_2_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_47_2_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(unc_47_2) <- librarySizeFactors(unc_47_2)
unc_47_2 <- normalize(unc_47_2)
unc_47_2
saveRDS(unc_47_2, "./final_gene_model/3131-ST-1/SoupX/020620_unc_47_2_sce_uncorrected.rds")

sizeFactors(unc_47_2_corrected) <- librarySizeFactors(unc_47_2_corrected)
unc_47_2_corrected <- normalize(unc_47_2_corrected)
unc_47_2_corrected
saveRDS(unc_47_2_corrected, "./final_gene_model/3131-ST-1/SoupX/020620_unc_47_2_sce_new_SoupX_corrected.rds")

# 2966-ST-1

my_dir <- "./final_gene_model/2966-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

nmr_1 <- read10xCounts(my_dir)
colnames(nmr_1) <- nmr_1$Barcode
nmr_1
head(rowData(nmr_1))
rowData(nmr_1) <- rowData(nmr_1)[,1:2]

rowData(nmr_1)$Symbol <- as.character(vlookup(rowData(nmr_1)$ID, gene_list))
head(rowData(nmr_1))

bcrank <- barcodeRanks(counts(nmr_1))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(nmr_1), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  1118    2644  733518   

c.keep <- defaultDrops(counts(nmr_1), expected=5000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
##  736875         58        347

nmr_1$Detection <- detection
nmr_1$PValue <- e.out$PValue

nmr_1 <- nmr_1[, keep]
nmr_1
## class: SingleCellExperiment 
## dim: 46911 2702

# Creating a filtered matrix to use for SoupX corrections

nmr_1_filtered_counts <- counts(nmr_1)
writeMM(nmr_1_filtered_counts, "./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(nmr_1_filtered_counts, "./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(nmr_1)
write.table(fil.barcodes, "./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/2966-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(nmr_1) <- librarySizeFactors(nmr_1)
nmr_1 <- normalize(nmr_1)
nmr_1S <- as.Seurat(nmr_1)

nmr_1S <- NormalizeData(nmr_1S)
nmr_1S <- FindVariableFeatures(nmr_1S, selection.method = "vst", nfeatures = 4000)
nmr_1S <- ScaleData(nmr_1S)
nmr_1S <- RunPCA(nmr_1S, features = VariableFeatures(nmr_1S), npcs = 125)

ElbowPlot(nmr_1S, ndims = 125)

nmr_1S <- FindNeighbors(nmr_1S, dims = 1:120)
nmr_1S <- FindClusters(nmr_1S, resolution = 1.2)

nmr_1S <- RunUMAP(nmr_1S, dims = 1:120)
DimPlot(nmr_1S, reduction = "umap")
FeaturePlot(nmr_1S, features = "WBGene00001461")
FeaturePlot(nmr_1S, features = "WBGene00001463")

umap.nmr_1 <- nmr_1S@reductions$umap

nmr_1_DR <- as.data.frame(Embeddings(umap.nmr_1))
head(Idents(nmr_1S))
nmr_1_DR$Cluster <- Idents(nmr_1S)
head(nmr_1_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/2966-ST-1/SoupX"
sclnmr_1 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclnmr_1$toc[1:5,1:5]
head(rownames(nmr_1_DR))

rownames(nmr_1_DR) <- substr(rownames(nmr_1_DR), 1, 16)
head(nmr_1_DR)
nmr_1_DR$flp20 <- sclnmr_1$toc["WBGene00001463", rownames(nmr_1_DR)]
gg <- ggplot(nmr_1_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = flp20 > 0))
plot(gg)

rownames(sclnmr_1$toc) <- as.character(vlookup(rownames(sclnmr_1$toc), gene_list))
sclnmr_1$toc[1:5,1:5]
rownames(sclnmr_1$soupProfile) <- rownames(sclnmr_1$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclnmr_1, "flp-20", nmr_1_DR)
plot(gg)
plotMarkerMap(sclnmr_1, "flp-19", nmr_1_DR)
plotMarkerMap(sclnmr_1, "flp-9")
plotMarkerMap(sclnmr_1, "snet-1")
plotMarkerMap(sclnmr_1, geneSet = c("flp-19", "flp-9", "flp-14"), nmr_1_DR)
head(sclnmr_1$soupProfile[order(sclnmr_1$soupProfile$est, decreasing = TRUE), ], n = 20)

# Now it works.

head(nmr_1_DR)
sclnmr_1 <- setDR(sclnmr_1, nmr_1_DR)

sclnmr_1 <- setClusters(sclnmr_1, nmr_1_DR$Cluster)
plotMarkerMap(sclnmr_1, "nlp-21")
plotMarkerMap(sclnmr_1, "flp-14")
plotMarkerDistribution(sclnmr_1)
nmr_1.genes <- c("flp-9", "snet-1")

useToEst = estimateNonExpressingCells(sclnmr_1, nonExpressedGeneList = list(nmr_1 = nmr_1.genes))
plotMarkerMap(sclnmr_1, geneSet = nmr_1.genes, DR = nmr_1_DR, useToEst = useToEst)

sclnmr_1 <- calculateContaminationFraction(sclnmr_1, list(nmr_1 = nmr_1.genes), useToEst = useToEst)
# Estimated global contamination fraction of 7.83%

system.time(out.nmr_1 <- adjustCounts(sclnmr_1, roundToInt = TRUE))
system.time(out.nmr_1.ni <- adjustCounts(sclnmr_1))

cntSoggy = rowSums(sclnmr_1$toc > 0)
cntStrained = rowSums(out.nmr_1 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclnmr_1$toc > out.nmr_1)/rowSums(sclnmr_1$toc > 0)), n = 20)

plotChangeMap(sclnmr_1, out.nmr_1, "flp-9")
plotChangeMap(sclnmr_1, out.nmr_1, "snet-1")
plotChangeMap(sclnmr_1, out.nmr_1, "flp-19")
plotChangeMap(sclnmr_1, out.nmr_1, "flp-18")
plotChangeMap(sclnmr_1, out.nmr_1, "glr-1")
plotChangeMap(sclnmr_1, out.nmr_1, "nmr-2")
plotChangeMap(sclnmr_1, out.nmr_1, "nduo-6")

out.nmr_1[1:5,1:5]
sclnmr_1$toc[1:5,1:5]

saveRDS(sclnmr_1, "./final_gene_model/2966-ST-1/SoupX/soupX_object_nmr_1.rds")
saveRDS(out.nmr_1, "./final_gene_model/2966-ST-1/SoupX/soupX_corrected_matrix_nmr_1.rds")
saveRDS(out.nmr_1.ni, "./final_gene_model/2966-ST-1/SoupX/soupX_corrected_matrix_nmr_1_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.nmr_1) <- paste(colnames(out.nmr_1), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.nmr_1) <- as.character(vlookup(rownames(out.nmr_1), gene_list, result_column = 1, lookup_column = 2))
out.nmr_1[1:5,1:5]
head(colData(nmr_1))

nmr_1_corrected <- SingleCellExperiment(assays = list(counts = out.nmr_1), colData = colData(nmr_1))  
nmr_1
nmr_1_corrected
head(rowData(nmr_1))
head(rowData(nmr_1_corrected))
rowData(nmr_1_corrected) <- rowData(nmr_1)
nmr_1_corrected

nmr_1 <- calculateQCMetrics(nmr_1, feature_controls = list(Mito = which(rowData(nmr_1)$ID == "WBGene00010959" | rowData(nmr_1)$ID == "WBGene00010961"| rowData(nmr_1)$ID == "WBGene00010966" | rowData(nmr_1)$ID == "WBGene00010963" | rowData(nmr_1)$ID =="WBGene00010957" | rowData(nmr_1)$ID == "WBGene00010967" | rowData(nmr_1)$ID == "WBGene00010958" | rowData(nmr_1)$ID == "WBGene00010960" | rowData(nmr_1)$ID == "WBGene00010962" | rowData(nmr_1)$ID == "WBGene00000829" | rowData(nmr_1)$ID =="WBGene00010965" | rowData(nmr_1)$ID == "WBGene00010964")))
nmr_1_corrected <- calculateQCMetrics(nmr_1_corrected, feature_controls = list(Mito = which(rowData(nmr_1_corrected)$ID == "WBGene00010959" | rowData(nmr_1_corrected)$ID == "WBGene00010961"| rowData(nmr_1_corrected)$ID == "WBGene00010966" | rowData(nmr_1_corrected)$ID == "WBGene00010963" | rowData(nmr_1_corrected)$ID =="WBGene00010957" | rowData(nmr_1_corrected)$ID == "WBGene00010967" | rowData(nmr_1_corrected)$ID == "WBGene00010958" | rowData(nmr_1_corrected)$ID == "WBGene00010960" | rowData(nmr_1_corrected)$ID == "WBGene00010962" | rowData(nmr_1_corrected)$ID == "WBGene00000829" | rowData(nmr_1_corrected)$ID =="WBGene00010965" | rowData(nmr_1_corrected)$ID == "WBGene00010964")))

summary(nmr_1$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   2.052   3.134   5.088   5.172  98.344

summary(nmr_1_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.020   1.884   3.969   3.897  98.826 

table(nmr_1$pct_counts_Mito < 20)
## FALSE  TRUE
##   91  2611

table(nmr_1_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##   81  2621 

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(nmr_1$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(nmr_1$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(nmr_1$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(nmr_1_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(nmr_1_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(nmr_1_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(nmr_1) <- librarySizeFactors(nmr_1)
nmr_1 <- normalize(nmr_1)
nmr_1
saveRDS(nmr_1, "./final_gene_model/2966-ST-1/SoupX/020620_nmr_1_sce_uncorrected.rds")

sizeFactors(nmr_1_corrected) <- librarySizeFactors(nmr_1_corrected)
nmr_1_corrected <- normalize(nmr_1_corrected)
nmr_1_corrected
saveRDS(nmr_1_corrected, "./final_gene_model/2966-ST-1/SoupX/020620_nmr_1_sce_new_SoupX_corrected.rds")

# 3070-ST-1

my_dir <- "./final_gene_model/3070-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

eat_4 <- read10xCounts(my_dir)
colnames(eat_4) <- eat_4$Barcode
eat_4
head(rowData(eat_4))
rowData(eat_4) <- rowData(eat_4)[,1:2]

rowData(eat_4)$Symbol <- as.character(vlookup(rowData(eat_4)$ID, gene_list))
head(rowData(eat_4))

bcrank <- barcodeRanks(counts(eat_4))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(eat_4), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  23424   14615  699241  

c.keep <- defaultDrops(counts(eat_4), expected=5000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
##  730365        201       6714 

eat_4$Detection <- detection
eat_4$PValue <- e.out$PValue

eat_4 <- eat_4[, keep]
eat_4
## class: SingleCellExperiment 
## dim: 46911 14816

# Creating a filtered matrix to use for SoupX corrections

eat_4_filtered_counts <- counts(eat_4)
writeMM(eat_4_filtered_counts, "./final_gene_model/3070-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3070-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(eat_4_filtered_counts, "./final_gene_model/3070-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(eat_4)
write.table(fil.barcodes, "./final_gene_model/3070-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3070-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3070-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(eat_4) <- librarySizeFactors(eat_4)
eat_4 <- normalize(eat_4)
eat_4S <- as.Seurat(eat_4)

eat_4S <- NormalizeData(eat_4S)
eat_4S <- FindVariableFeatures(eat_4S, selection.method = "vst", nfeatures = 4000)
eat_4S <- ScaleData(eat_4S)
eat_4S <- RunPCA(eat_4S, features = VariableFeatures(eat_4S), npcs = 145)

ElbowPlot(eat_4S, ndims = 145)

eat_4S <- FindNeighbors(eat_4S, dims = 1:120)
eat_4S <- FindClusters(eat_4S, resolution = 1.2)

eat_4S <- RunUMAP(eat_4S, dims = 1:120)
DimPlot(eat_4S, reduction = "umap")
FeaturePlot(eat_4S, features = "WBGene00000845")
FeaturePlot(eat_4S, features = "WBGene00018877")

umap.eat_4 <- eat_4S@reductions$umap

eat_4_DR <- as.data.frame(Embeddings(umap.eat_4))
head(Idents(eat_4S))
eat_4_DR$Cluster <- Idents(eat_4S)
head(eat_4_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3070-ST-1/SoupX"
scleat_4 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

scleat_4$toc[1:5,1:5]
head(rownames(eat_4_DR))

rownames(eat_4_DR) <- substr(rownames(eat_4_DR), 1, 16)
head(eat_4_DR)
eat_4_DR$aman1 <- scleat_4$toc["WBGene00018877", rownames(eat_4_DR)]
gg <- ggplot(eat_4_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(scleat_4$toc) <- as.character(vlookup(rownames(scleat_4$toc), gene_list))
scleat_4$toc[1:5,1:5]
rownames(scleat_4$soupProfile) <- rownames(scleat_4$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(scleat_4, "aman-1", eat_4_DR)
plot(gg)
plotMarkerMap(scleat_4, "cup-4", eat_4_DR)
plotMarkerMap(scleat_4, "pat-10", eat_4_DR)
plotMarkerMap(scleat_4, "cpn-3", eat_4_DR)
plotMarkerMap(scleat_4, geneSet = c("cup-4", "aman-1", "pat-10", "cpn-3"), eat_4_DR)
head(scleat_4$soupProfile[order(scleat_4$soupProfile$est, decreasing = TRUE), ], n = 20)

head(eat_4_DR)
scleat_4 <- setDR(scleat_4, eat_4_DR)

scleat_4 <- setClusters(scleat_4, eat_4_DR$Cluster)
plotMarkerMap(scleat_4, "nlp-17")
plotMarkerMap(scleat_4, "far-1")
plotMarkerDistribution(scleat_4)

eat_4.genes <- c("far-1", "cup-4", "aman-1", "pat-10", "cpn-3", "Y73F4A.1", "nlp-17", "mec-17")
useToEst = estimateNonExpressingCells(scleat_4, nonExpressedGeneList = list(eat_4 = eat_4.genes))
plotMarkerMap(scleat_4, geneSet = eat_4.genes, DR = eat_4_DR, useToEst = useToEst)
scleat_4 <- calculateContaminationFraction(scleat_4, list(eat_4 = eat_4.genes), useToEst = useToEst)
# Estimated global contamination fraction of 10.61%

system.time(out.eat_4 <- adjustCounts(scleat_4, roundToInt = TRUE))
system.time(out.eat_4.ni <- adjustCounts(scleat_4))

cntSoggy = rowSums(scleat_4$toc > 0)
cntStrained = rowSums(out.eat_4 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(scleat_4$toc > out.eat_4)/rowSums(scleat_4$toc > 0)), n = 20)

plotChangeMap(scleat_4, out.eat_4, "cup-4")
plotChangeMap(scleat_4, out.eat_4, "aman-1")
plotChangeMap(scleat_4, out.eat_4, "mec-17")
plotChangeMap(scleat_4, out.eat_4, "far-1")
plotChangeMap(scleat_4, out.eat_4, "mlc-3")
plotChangeMap(scleat_4, out.eat_4, "cpn-3")
plotChangeMap(scleat_4, out.eat_4, "nlp-17")

out.eat_4[1:5,1:5]
scleat_4$toc[1:5,1:5]

saveRDS(scleat_4, "./final_gene_model/3070-ST-1/SoupX/soupX_object_eat_4.rds")
saveRDS(out.eat_4, "./final_gene_model/3070-ST-1/SoupX/soupX_corrected_matrix_eat_4.rds")
saveRDS(out.eat_4.ni, "./final_gene_model/3070-ST-1/SoupX/soupX_corrected_matrix_eat_4_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.eat_4) <- paste(colnames(out.eat_4), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.eat_4) <- as.character(vlookup(rownames(out.eat_4), gene_list, result_column = 1, lookup_column = 2))
out.eat_4[1:5,1:5]
head(colData(eat_4))

eat_4_corrected <- SingleCellExperiment(assays = list(counts = out.eat_4), colData = colData(eat_4))  
eat_4
eat_4_corrected
head(rowData(eat_4))
head(rowData(eat_4_corrected))
rowData(eat_4_corrected) <- rowData(eat_4)
eat_4_corrected

eat_4 <- calculateQCMetrics(eat_4, feature_controls = list(Mito = which(rowData(eat_4)$ID == "WBGene00010959" | rowData(eat_4)$ID == "WBGene00010961"| rowData(eat_4)$ID == "WBGene00010966" | rowData(eat_4)$ID == "WBGene00010963" | rowData(eat_4)$ID =="WBGene00010957" | rowData(eat_4)$ID == "WBGene00010967" | rowData(eat_4)$ID == "WBGene00010958" | rowData(eat_4)$ID == "WBGene00010960" | rowData(eat_4)$ID == "WBGene00010962" | rowData(eat_4)$ID == "WBGene00000829" | rowData(eat_4)$ID =="WBGene00010965" | rowData(eat_4)$ID == "WBGene00010964")))
eat_4_corrected <- calculateQCMetrics(eat_4_corrected, feature_controls = list(Mito = which(rowData(eat_4_corrected)$ID == "WBGene00010959" | rowData(eat_4_corrected)$ID == "WBGene00010961"| rowData(eat_4_corrected)$ID == "WBGene00010966" | rowData(eat_4_corrected)$ID == "WBGene00010963" | rowData(eat_4_corrected)$ID =="WBGene00010957" | rowData(eat_4_corrected)$ID == "WBGene00010967" | rowData(eat_4_corrected)$ID == "WBGene00010958" | rowData(eat_4_corrected)$ID == "WBGene00010960" | rowData(eat_4_corrected)$ID == "WBGene00010962" | rowData(eat_4_corrected)$ID == "WBGene00000829" | rowData(eat_4_corrected)$ID =="WBGene00010965" | rowData(eat_4_corrected)$ID == "WBGene00010964")))

summary(eat_4$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   2.406   3.704   5.380   6.285  96.755 

summary(eat_4_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.621   2.993   4.814   5.673  97.065

table(eat_4$pct_counts_Mito < 20)
## FALSE  TRUE
##   396 14420

table(eat_4_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  431 14385 

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(eat_4$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(eat_4$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(eat_4$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(eat_4_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(eat_4_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(eat_4_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(eat_4) <- librarySizeFactors(eat_4)
eat_4 <- normalize(eat_4)
eat_4
saveRDS(eat_4, "./final_gene_model/3070-ST-1/SoupX/020620_eat_4_sce_uncorrected.rds")

sizeFactors(eat_4_corrected) <- librarySizeFactors(eat_4_corrected)
eat_4_corrected <- normalize(eat_4_corrected)
eat_4_corrected
saveRDS(eat_4_corrected, "./final_gene_model/3070-ST-1/SoupX/020620_eat_4_sce_new_SoupX_corrected.rds")

# 3183-ST-1

my_dir <- "./final_gene_model/3183-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

unc_3 <- read10xCounts(my_dir)
colnames(unc_3) <- unc_3$Barcode
unc_3
head(rowData(unc_3))
rowData(unc_3) <- rowData(unc_3)[,1:2]

rowData(unc_3)$Symbol <- as.character(vlookup(rowData(unc_3)$ID, gene_list))
head(rowData(unc_3))

bcrank <- barcodeRanks(counts(unc_3))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(unc_3), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  6792    7132  723356   

c.keep <- defaultDrops(counts(unc_3), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
##  734278         13       2989

unc_3$Detection <- detection
unc_3$PValue <- e.out$PValue

unc_3 <- unc_3[, keep]
unc_3
## class: SingleCellExperiment 
## dim: 46911 7145

# Creating a filtered matrix to use for SoupX corrections

unc_3_filtered_counts <- counts(unc_3)
writeMM(unc_3_filtered_counts, "./final_gene_model/3183-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3183-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(unc_3_filtered_counts, "./final_gene_model/3183-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(unc_3)
write.table(fil.barcodes, "./final_gene_model/3183-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3183-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3183-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(unc_3) <- librarySizeFactors(unc_3)
unc_3 <- normalize(unc_3)
unc_3S <- as.Seurat(unc_3)

unc_3S <- NormalizeData(unc_3S)
unc_3S <- FindVariableFeatures(unc_3S, selection.method = "vst", nfeatures = 4000)
unc_3S <- ScaleData(unc_3S)
unc_3S <- RunPCA(unc_3S, features = VariableFeatures(unc_3S), npcs = 145)

ElbowPlot(unc_3S, ndims = 145)

unc_3S <- FindNeighbors(unc_3S, dims = 1:120)
unc_3S <- FindClusters(unc_3S, resolution = 1.2)

unc_3S <- RunUMAP(unc_3S, dims = 1:120)
DimPlot(unc_3S, reduction = "umap")
FeaturePlot(unc_3S, features = "WBGene00000845")
FeaturePlot(unc_3S, features = "WBGene00018877")

umap.unc_3 <- unc_3S@reductions$umap

unc_3_DR <- as.data.frame(Embeddings(umap.unc_3))
head(Idents(unc_3S))
unc_3_DR$Cluster <- Idents(unc_3S)
head(unc_3_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3183-ST-1/SoupX"
sclunc_3 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclunc_3$toc[1:5,1:5]
head(rownames(unc_3_DR))

rownames(unc_3_DR) <- substr(rownames(unc_3_DR), 1, 16)
head(unc_3_DR)
unc_3_DR$aman1 <- sclunc_3$toc["WBGene00018877", rownames(unc_3_DR)]
gg <- ggplot(unc_3_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(sclunc_3$toc) <- as.character(vlookup(rownames(sclunc_3$toc), gene_list))
sclunc_3$toc[1:5,1:5]
rownames(sclunc_3$soupProfile) <- rownames(sclunc_3$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclunc_3, "aman-1", unc_3_DR)
plot(gg)
plotMarkerMap(sclunc_3, "cup-4", unc_3_DR)
plotMarkerMap(sclunc_3, "pat-10", unc_3_DR)
plotMarkerMap(sclunc_3, "cpn-3", unc_3_DR)
plotMarkerMap(sclunc_3, geneSet = c("cup-4", "aman-1", "pat-10", "cpn-3"), unc_3_DR)
head(sclunc_3$soupProfile[order(sclunc_3$soupProfile$est, decreasing = TRUE), ], n = 20)

head(unc_3_DR)
sclunc_3 <- setDR(sclunc_3, unc_3_DR)

sclunc_3 <- setClusters(sclunc_3, unc_3_DR$Cluster)
plotMarkerMap(sclunc_3, "snet-1")
plotMarkerMap(sclunc_3, "flp-12")
plotMarkerMap(sclunc_3, "far-1")
plotMarkerDistribution(sclunc_3)

unc_3.genes <- c("far-1", "cup-4", "aman-1", "pat-10", "cpn-3", "ttr-1", "Y73F4A.1", "flp-12", "snet-1")
useToEst = estimateNonExpressingCells(sclunc_3, nonExpressedGeneList = list(unc_3 = unc_3.genes))
plotMarkerMap(sclunc_3, geneSet = unc_3.genes, DR = unc_3_DR, useToEst = useToEst)
sclunc_3 <- calculateContaminationFraction(sclunc_3, list(unc_3 = unc_3.genes), useToEst = useToEst)
# Estimated global contamination fraction of 6.52%

system.time(out.unc_3 <- adjustCounts(sclunc_3, roundToInt = TRUE))
system.time(out.unc_3.ni <- adjustCounts(sclunc_3))

cntSoggy = rowSums(sclunc_3$toc > 0)
cntStrained = rowSums(out.unc_3 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclunc_3$toc > out.unc_3)/rowSums(sclunc_3$toc > 0)), n = 20)

plotChangeMap(sclunc_3, out.unc_3, "cup-4")
plotChangeMap(sclunc_3, out.unc_3, "flp-12")
plotChangeMap(sclunc_3, out.unc_3, "flp-11")
plotChangeMap(sclunc_3, out.unc_3, "ilys-4")
plotChangeMap(sclunc_3, out.unc_3, "unc-25")
plotChangeMap(sclunc_3, out.unc_3, "K09F6.13")
plotChangeMap(sclunc_3, out.unc_3, "nduo-6")

out.unc_3[1:5,1:5]
sclunc_3$toc[1:5,1:5]

saveRDS(sclunc_3, "./final_gene_model/3183-ST-1/SoupX/soupX_object_unc_3.rds")
saveRDS(out.unc_3, "./final_gene_model/3183-ST-1/SoupX/soupX_corrected_matrix_unc_3.rds")
saveRDS(out.unc_3.ni, "./final_gene_model/3183-ST-1/SoupX/soupX_corrected_matrix_unc_3_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.unc_3) <- paste(colnames(out.unc_3), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.unc_3) <- as.character(vlookup(rownames(out.unc_3), gene_list, result_column = 1, lookup_column = 2))
out.unc_3[1:5,1:5]
head(colData(unc_3))

unc_3_corrected <- SingleCellExperiment(assays = list(counts = out.unc_3), colData = colData(unc_3))  
unc_3
unc_3_corrected
head(rowData(unc_3))
head(rowData(unc_3_corrected))
rowData(unc_3_corrected) <- rowData(unc_3)
unc_3_corrected

unc_3 <- calculateQCMetrics(unc_3, feature_controls = list(Mito = which(rowData(unc_3)$ID == "WBGene00010959" | rowData(unc_3)$ID == "WBGene00010961"| rowData(unc_3)$ID == "WBGene00010966" | rowData(unc_3)$ID == "WBGene00010963" | rowData(unc_3)$ID =="WBGene00010957" | rowData(unc_3)$ID == "WBGene00010967" | rowData(unc_3)$ID == "WBGene00010958" | rowData(unc_3)$ID == "WBGene00010960" | rowData(unc_3)$ID == "WBGene00010962" | rowData(unc_3)$ID == "WBGene00000829" | rowData(unc_3)$ID =="WBGene00010965" | rowData(unc_3)$ID == "WBGene00010964")))
unc_3_corrected <- calculateQCMetrics(unc_3_corrected, feature_controls = list(Mito = which(rowData(unc_3_corrected)$ID == "WBGene00010959" | rowData(unc_3_corrected)$ID == "WBGene00010961"| rowData(unc_3_corrected)$ID == "WBGene00010966" | rowData(unc_3_corrected)$ID == "WBGene00010963" | rowData(unc_3_corrected)$ID =="WBGene00010957" | rowData(unc_3_corrected)$ID == "WBGene00010967" | rowData(unc_3_corrected)$ID == "WBGene00010958" | rowData(unc_3_corrected)$ID == "WBGene00010960" | rowData(unc_3_corrected)$ID == "WBGene00010962" | rowData(unc_3_corrected)$ID == "WBGene00000829" | rowData(unc_3_corrected)$ID =="WBGene00010965" | rowData(unc_3_corrected)$ID == "WBGene00010964")))

summary(unc_3$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.132   1.795   3.490   3.333  48.256 

summary(unc_3_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.6536  1.3024  3.0483  2.7500 54.3396  

table(unc_3$pct_counts_Mito < 20)
## FALSE  TRUE
##   168  6977 

table(unc_3_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  172  6973 

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(unc_3$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_3$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_3$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(unc_3_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_3_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_3_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(unc_3) <- librarySizeFactors(unc_3)
unc_3 <- normalize(unc_3)
unc_3
saveRDS(unc_3, "./final_gene_model/3183-ST-1/SoupX/020620_unc_3_sce_uncorrected.rds")

sizeFactors(unc_3_corrected) <- librarySizeFactors(unc_3_corrected)
unc_3_corrected <- normalize(unc_3_corrected)
unc_3_corrected
saveRDS(unc_3_corrected, "./final_gene_model/3183-ST-1/SoupX/020620_unc_3_sce_new_SoupX_corrected.rds")

# 3239-ST-1

my_dir <- "./final_gene_model/3239-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

ift_20 <- read10xCounts(my_dir)
colnames(ift_20) <- ift_20$Barcode
ift_20
head(rowData(ift_20))
rowData(ift_20) <- rowData(ift_20)[,1:2]

rowData(ift_20)$Symbol <- as.character(vlookup(rowData(ift_20)$ID, gene_list))
head(rowData(ift_20))

bcrank <- barcodeRanks(counts(ift_20))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(ift_20), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  558    4346  732376   

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability")

c.keep <- defaultDrops(counts(ift_20), expected=5000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
##  736384         30        866 

ift_20$Detection <- detection
ift_20$PValue <- e.out$PValue

ift_20 <- ift_20[, keep]
ift_20
## class: SingleCellExperiment 
## dim: 46911 4376

# Creating a filtered matrix to use for SoupX corrections

ift_20_filtered_counts <- counts(ift_20)
writeMM(ift_20_filtered_counts, "./final_gene_model/3239-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3239-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(ift_20_filtered_counts, "./final_gene_model/3239-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(ift_20)
write.table(fil.barcodes, "./final_gene_model/3239-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3239-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3239-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(ift_20) <- librarySizeFactors(ift_20)
ift_20 <- normalize(ift_20)
ift_20S <- as.Seurat(ift_20)

ift_20S <- NormalizeData(ift_20S)
ift_20S <- FindVariableFeatures(ift_20S, selection.method = "vst", nfeatures = 4000)
ift_20S <- ScaleData(ift_20S)
ift_20S <- RunPCA(ift_20S, features = VariableFeatures(ift_20S), npcs = 145)

ElbowPlot(ift_20S, ndims = 145)

ift_20S <- FindNeighbors(ift_20S, dims = 1:120)
ift_20S <- FindClusters(ift_20S, resolution = 1.2)

ift_20S <- RunUMAP(ift_20S, dims = 1:120)
DimPlot(ift_20S, reduction = "umap")
FeaturePlot(ift_20S, features = "WBGene00016114")

umap.ift_20 <- ift_20S@reductions$umap

ift_20_DR <- as.data.frame(Embeddings(umap.ift_20))
head(Idents(ift_20S))
ift_20_DR$Cluster <- Idents(ift_20S)
head(ift_20_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3239-ST-1/SoupX"
sclift_20 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclift_20$toc[1:5,1:5]
head(rownames(ift_20_DR))

rownames(ift_20_DR) <- substr(rownames(ift_20_DR), 1, 16)
head(ift_20_DR)
ift_20_DR$aman1 <- sclift_20$toc["WBGene00018877", rownames(ift_20_DR)]
gg <- ggplot(ift_20_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(sclift_20$toc) <- as.character(vlookup(rownames(sclift_20$toc), gene_list))
sclift_20$toc[1:5,1:5]
rownames(sclift_20$soupProfile) <- rownames(sclift_20$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclift_20, "flp-27", ift_20_DR)
plot(gg)
plotMarkerMap(sclift_20, "flp-9", ift_20_DR)
plotMarkerMap(sclift_20, "daf-7", ift_20_DR)
plotMarkerMap(sclift_20, "ins-26", ift_20_DR)
plotMarkerMap(sclift_20, geneSet = c("R102.2", "ins-30", "flp-27", "ins-3"), ift_20_DR)
head(sclift_20$soupProfile[order(sclift_20$soupProfile$est, decreasing = TRUE), ], n = 20)

head(ift_20_DR)
sclift_20 <- setDR(sclift_20, ift_20_DR)

sclift_20 <- setClusters(sclift_20, ift_20_DR$Cluster)
plotMarkerMap(sclift_20, "snet-1")
plotMarkerMap(sclift_20, "C33A12.4")
plotMarkerDistribution(sclift_20)

ift_20.genes <- c("R102.2", "ins-30", "flp-27", "ins-3", "ins-6")
useToEst = estimateNonExpressingCells(sclift_20, nonExpressedGeneList = list(ift_20 = ift_20.genes))
plotMarkerMap(sclift_20, geneSet = ift_20.genes, DR = ift_20_DR, useToEst = useToEst)
sclift_20 <- calculateContaminationFraction(sclift_20, list(ift_20 = ift_20.genes), useToEst = useToEst)
# Estimated global contamination fraction of 11.23%

system.time(out.ift_20 <- adjustCounts(sclift_20, roundToInt = TRUE))
system.time(out.ift_20.ni <- adjustCounts(sclift_20))

cntSoggy = rowSums(sclift_20$toc > 0)
cntStrained = rowSums(out.ift_20 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclift_20$toc > out.ift_20)/rowSums(sclift_20$toc > 0)), n = 20)

plotChangeMap(sclift_20, out.ift_20, "flp-27")
plotChangeMap(sclift_20, out.ift_20, "ins-30")
plotChangeMap(sclift_20, out.ift_20, "ins-3")
plotChangeMap(sclift_20, out.ift_20, "ins-6")
plotChangeMap(sclift_20, out.ift_20, "unc-25")
plotChangeMap(sclift_20, out.ift_20, "K09F6.13")
plotChangeMap(sclift_20, out.ift_20, "nduo-6")

out.ift_20[1:5,1:5]
sclift_20$toc[1:5,1:5]

saveRDS(sclift_20, "./final_gene_model/3239-ST-1/SoupX/soupX_object_ift_20.rds")
saveRDS(out.ift_20, "./final_gene_model/3239-ST-1/SoupX/soupX_corrected_matrix_ift_20.rds")
saveRDS(out.ift_20.ni, "./final_gene_model/3239-ST-1/SoupX/soupX_corrected_matrix_ift_20_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.ift_20) <- paste(colnames(out.ift_20), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.ift_20) <- as.character(vlookup(rownames(out.ift_20), gene_list, result_column = 1, lookup_column = 2))
out.ift_20[1:5,1:5]
head(colData(ift_20))

ift_20_corrected <- SingleCellExperiment(assays = list(counts = out.ift_20), colData = colData(ift_20))  
ift_20
ift_20_corrected
head(rowData(ift_20))
head(rowData(ift_20_corrected))
rowData(ift_20_corrected) <- rowData(ift_20)
ift_20_corrected

ift_20 <- calculateQCMetrics(ift_20, feature_controls = list(Mito = which(rowData(ift_20)$ID == "WBGene00010959" | rowData(ift_20)$ID == "WBGene00010961"| rowData(ift_20)$ID == "WBGene00010966" | rowData(ift_20)$ID == "WBGene00010963" | rowData(ift_20)$ID =="WBGene00010957" | rowData(ift_20)$ID == "WBGene00010967" | rowData(ift_20)$ID == "WBGene00010958" | rowData(ift_20)$ID == "WBGene00010960" | rowData(ift_20)$ID == "WBGene00010962" | rowData(ift_20)$ID == "WBGene00000829" | rowData(ift_20)$ID =="WBGene00010965" | rowData(ift_20)$ID == "WBGene00010964")))
ift_20_corrected <- calculateQCMetrics(ift_20_corrected, feature_controls = list(Mito = which(rowData(ift_20_corrected)$ID == "WBGene00010959" | rowData(ift_20_corrected)$ID == "WBGene00010961"| rowData(ift_20_corrected)$ID == "WBGene00010966" | rowData(ift_20_corrected)$ID == "WBGene00010963" | rowData(ift_20_corrected)$ID =="WBGene00010957" | rowData(ift_20_corrected)$ID == "WBGene00010967" | rowData(ift_20_corrected)$ID == "WBGene00010958" | rowData(ift_20_corrected)$ID == "WBGene00010960" | rowData(ift_20_corrected)$ID == "WBGene00010962" | rowData(ift_20_corrected)$ID == "WBGene00000829" | rowData(ift_20_corrected)$ID =="WBGene00010965" | rowData(ift_20_corrected)$ID == "WBGene00010964")))

summary(ift_20$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.8262  1.3782  2.6566  2.5300 56.5574 

summary(ift_20_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.00000  0.00000  0.04969  1.20422  0.63768 59.78261 

table(ift_20$pct_counts_Mito < 20)
## FALSE  TRUE
##   75  4301 

table(ift_20_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  56  4320  

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(ift_20$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(ift_20$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(ift_20$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(ift_20_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(ift_20_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(ift_20_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(ift_20) <- librarySizeFactors(ift_20)
ift_20 <- normalize(ift_20)
ift_20
saveRDS(ift_20, "./final_gene_model/3239-ST-1/SoupX/020620_ift_20_sce_uncorrected.rds")

sizeFactors(ift_20_corrected) <- librarySizeFactors(ift_20_corrected)
ift_20_corrected <- normalize(ift_20_corrected)
ift_20_corrected
saveRDS(ift_20_corrected, "./final_gene_model/3239-ST-1/SoupX/020620_ift_20_sce_new_SoupX_corrected.rds")

# 3441-ST-1

my_dir <- "./final_gene_model/3441-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

cho_1_1 <- read10xCounts(my_dir)
colnames(cho_1_1) <- cho_1_1$Barcode
cho_1_1
head(rowData(cho_1_1))
rowData(cho_1_1) <- rowData(cho_1_1)[,1:2]

rowData(cho_1_1)$Symbol <- as.character(vlookup(rowData(cho_1_1)$ID, gene_list))
head(rowData(cho_1_1))

bcrank <- barcodeRanks(counts(cho_1_1))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(cho_1_1), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  4326    5357  727597  

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability")

c.keep <- defaultDrops(counts(cho_1_1), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
## 735143         50       2087 

cho_1_1$Detection <- detection
cho_1_1$PValue <- e.out$PValue

cho_1_1 <- cho_1_1[, keep]
cho_1_1
## class: SingleCellExperiment 
## dim: 46911 5407

# Creating a filtered matrix to use for SoupX corrections

cho_1_1_filtered_counts <- counts(cho_1_1)
writeMM(cho_1_1_filtered_counts, "./final_gene_model/3441-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3441-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(cho_1_1_filtered_counts, "./final_gene_model/3441-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(cho_1_1)
write.table(fil.barcodes, "./final_gene_model/3441-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3441-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3441-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(cho_1_1) <- librarySizeFactors(cho_1_1)
cho_1_1 <- normalize(cho_1_1)
cho_1_1S <- as.Seurat(cho_1_1)

cho_1_1S <- NormalizeData(cho_1_1S)
cho_1_1S <- FindVariableFeatures(cho_1_1S, selection.method = "vst", nfeatures = 4000)
cho_1_1S <- ScaleData(cho_1_1S)
cho_1_1S <- RunPCA(cho_1_1S, features = VariableFeatures(cho_1_1S), npcs = 145)

ElbowPlot(cho_1_1S, ndims = 145)

cho_1_1S <- FindNeighbors(cho_1_1S, dims = 1:135)
cho_1_1S <- FindClusters(cho_1_1S, resolution = 1.2)

cho_1_1S <- RunUMAP(cho_1_1S, dims = 1:135)
DimPlot(cho_1_1S, reduction = "umap")
FeaturePlot(cho_1_1S, features = "WBGene00016114")

umap.cho_1_1 <- cho_1_1S@reductions$umap

cho_1_1_DR <- as.data.frame(Embeddings(umap.cho_1_1))
head(Idents(cho_1_1S))
cho_1_1_DR$Cluster <- Idents(cho_1_1S)
head(cho_1_1_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3441-ST-1/SoupX"
sclcho_1_1 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclcho_1_1$toc[1:5,1:5]
head(rownames(cho_1_1_DR))

rownames(cho_1_1_DR) <- substr(rownames(cho_1_1_DR), 1, 16)
head(cho_1_1_DR)
cho_1_1_DR$aman1 <- sclcho_1_1$toc["WBGene00018877", rownames(cho_1_1_DR)]
gg <- ggplot(cho_1_1_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(sclcho_1_1$toc) <- as.character(vlookup(rownames(sclcho_1_1$toc), gene_list))
sclcho_1_1$toc[1:5,1:5]
rownames(sclcho_1_1$soupProfile) <- rownames(sclcho_1_1$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclcho_1_1, "flp-27", cho_1_1_DR)
plot(gg)
plotMarkerMap(sclcho_1_1, "pdf-1", cho_1_1_DR)
plotMarkerMap(sclcho_1_1, "flp-12", cho_1_1_DR)
plotMarkerMap(sclcho_1_1, "flp-9", cho_1_1_DR)
plotMarkerMap(sclcho_1_1, geneSet = c("R102.2", "ins-30", "flp-27", "ins-3"), cho_1_1_DR)
head(sclcho_1_1$soupProfile[order(sclcho_1_1$soupProfile$est, decreasing = TRUE), ], n = 20)

head(cho_1_1_DR)
sclcho_1_1 <- setDR(sclcho_1_1, cho_1_1_DR)

sclcho_1_1 <- setClusters(sclcho_1_1, cho_1_1_DR$Cluster)
plotMarkerMap(sclcho_1_1, "snet-1")
plotMarkerMap(sclcho_1_1, "pdf-1")
plotMarkerDistribution(sclcho_1_1)

cho_1_1.genes <- c("flp-12", "snet-1", "pdf-1")
useToEst = estimateNonExpressingCells(sclcho_1_1, nonExpressedGeneList = list(cho_1_1 = cho_1_1.genes))
plotMarkerMap(sclcho_1_1, geneSet = cho_1_1.genes, DR = cho_1_1_DR, useToEst = useToEst)
sclcho_1_1 <- calculateContaminationFraction(sclcho_1_1, list(cho_1_1 = cho_1_1.genes), useToEst = useToEst)
# Estimated global contamination fraction of 6.06%

system.time(out.cho_1_1 <- adjustCounts(sclcho_1_1, roundToInt = TRUE))
system.time(out.cho_1_1.ni <- adjustCounts(sclcho_1_1))

cntSoggy = rowSums(sclcho_1_1$toc > 0)
cntStrained = rowSums(out.cho_1_1 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclcho_1_1$toc > out.cho_1_1)/rowSums(sclcho_1_1$toc > 0)), n = 20)

plotChangeMap(sclcho_1_1, out.cho_1_1, "flp-12")
plotChangeMap(sclcho_1_1, out.cho_1_1, "flp-9")
plotChangeMap(sclcho_1_1, out.cho_1_1, "snet-1")
plotChangeMap(sclcho_1_1, out.cho_1_1, "pdf-1")
plotChangeMap(sclcho_1_1, out.cho_1_1, "unc-25")
plotChangeMap(sclcho_1_1, out.cho_1_1, "K09F6.13")
plotChangeMap(sclcho_1_1, out.cho_1_1, "nduo-6")

out.cho_1_1[1:5,1:5]
sclcho_1_1$toc[1:5,1:5]

saveRDS(sclcho_1_1, "./final_gene_model/3441-ST-1/SoupX/soupX_object_cho_1_1.rds")
saveRDS(out.cho_1_1, "./final_gene_model/3441-ST-1/SoupX/soupX_corrected_matrix_cho_1_1.rds")
saveRDS(out.cho_1_1.ni, "./final_gene_model/3441-ST-1/SoupX/soupX_corrected_matrix_cho_1_1_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.cho_1_1) <- paste(colnames(out.cho_1_1), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.cho_1_1) <- as.character(vlookup(rownames(out.cho_1_1), gene_list, result_column = 1, lookup_column = 2))
out.cho_1_1[1:5,1:5]
head(colData(cho_1_1))

cho_1_1_corrected <- SingleCellExperiment(assays = list(counts = out.cho_1_1), colData = colData(cho_1_1))  
cho_1_1
cho_1_1_corrected
head(rowData(cho_1_1))
head(rowData(cho_1_1_corrected))
rowData(cho_1_1_corrected) <- rowData(cho_1_1)
cho_1_1_corrected

cho_1_1 <- calculateQCMetrics(cho_1_1, feature_controls = list(Mito = which(rowData(cho_1_1)$ID == "WBGene00010959" | rowData(cho_1_1)$ID == "WBGene00010961"| rowData(cho_1_1)$ID == "WBGene00010966" | rowData(cho_1_1)$ID == "WBGene00010963" | rowData(cho_1_1)$ID =="WBGene00010957" | rowData(cho_1_1)$ID == "WBGene00010967" | rowData(cho_1_1)$ID == "WBGene00010958" | rowData(cho_1_1)$ID == "WBGene00010960" | rowData(cho_1_1)$ID == "WBGene00010962" | rowData(cho_1_1)$ID == "WBGene00000829" | rowData(cho_1_1)$ID =="WBGene00010965" | rowData(cho_1_1)$ID == "WBGene00010964")))
cho_1_1_corrected <- calculateQCMetrics(cho_1_1_corrected, feature_controls = list(Mito = which(rowData(cho_1_1_corrected)$ID == "WBGene00010959" | rowData(cho_1_1_corrected)$ID == "WBGene00010961"| rowData(cho_1_1_corrected)$ID == "WBGene00010966" | rowData(cho_1_1_corrected)$ID == "WBGene00010963" | rowData(cho_1_1_corrected)$ID =="WBGene00010957" | rowData(cho_1_1_corrected)$ID == "WBGene00010967" | rowData(cho_1_1_corrected)$ID == "WBGene00010958" | rowData(cho_1_1_corrected)$ID == "WBGene00010960" | rowData(cho_1_1_corrected)$ID == "WBGene00010962" | rowData(cho_1_1_corrected)$ID == "WBGene00000829" | rowData(cho_1_1_corrected)$ID =="WBGene00010965" | rowData(cho_1_1_corrected)$ID == "WBGene00010964")))

summary(cho_1_1$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.690   3.010   5.997   6.007  96.685 

summary(cho_1_1_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.2349  1.1527  3.9399  3.3881 96.9274 

table(cho_1_1$pct_counts_Mito < 20)
## FALSE  TRUE
##   356  5051 

table(cho_1_1_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  253  5154  

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(cho_1_1$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(cho_1_1$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(cho_1_1$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(cho_1_1_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(cho_1_1_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(cho_1_1_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(cho_1_1) <- librarySizeFactors(cho_1_1)
cho_1_1 <- normalize(cho_1_1)
cho_1_1
saveRDS(cho_1_1, "./final_gene_model/3441-ST-1/SoupX/020620_cho_1_1_sce_uncorrected.rds")

sizeFactors(cho_1_1_corrected) <- librarySizeFactors(cho_1_1_corrected)
cho_1_1_corrected <- normalize(cho_1_1_corrected)
cho_1_1_corrected
saveRDS(cho_1_1_corrected, "./final_gene_model/3441-ST-1/SoupX/020620_cho_1_1_sce_new_SoupX_corrected.rds")


# 3441-ST-2

my_dir <- "./final_gene_model/3441-ST-2/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

cho_1_2 <- read10xCounts(my_dir)
colnames(cho_1_2) <- cho_1_2$Barcode
cho_1_2
head(rowData(cho_1_2))
rowData(cho_1_2) <- rowData(cho_1_2)[,1:2]

rowData(cho_1_2)$Symbol <- as.character(vlookup(rowData(cho_1_2)$ID, gene_list))
head(rowData(cho_1_2))

bcrank <- barcodeRanks(counts(cho_1_2))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(cho_1_2), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  4846    4682  727752 

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability", xlim = c(0,1000), ylim = c(0,2000))

c.keep <- defaultDrops(counts(cho_1_2), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
## 735528         72       1680 

cho_1_2$Detection <- detection
cho_1_2$PValue <- e.out$PValue

cho_1_2 <- cho_1_2[, keep]
cho_1_2
## class: SingleCellExperiment 
## dim: 46911 4754

# Creating a filtered matrix to use for SoupX corrections

cho_1_2_filtered_counts <- counts(cho_1_2)
writeMM(cho_1_2_filtered_counts, "./final_gene_model/3441-ST-2/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3441-ST-2/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(cho_1_2_filtered_counts, "./final_gene_model/3441-ST-2/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(cho_1_2)
write.table(fil.barcodes, "./final_gene_model/3441-ST-2/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3441-ST-2/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3441-ST-2/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(cho_1_2) <- librarySizeFactors(cho_1_2)
cho_1_2 <- normalize(cho_1_2)
cho_1_2S <- as.Seurat(cho_1_2)

cho_1_2S <- NormalizeData(cho_1_2S)
cho_1_2S <- FindVariableFeatures(cho_1_2S, selection.method = "vst", nfeatures = 4000)
cho_1_2S <- ScaleData(cho_1_2S)
cho_1_2S <- RunPCA(cho_1_2S, features = VariableFeatures(cho_1_2S), npcs = 145)

ElbowPlot(cho_1_2S, ndims = 145)

cho_1_2S <- FindNeighbors(cho_1_2S, dims = 1:125)
cho_1_2S <- FindClusters(cho_1_2S, resolution = 1.2)

cho_1_2S <- RunUMAP(cho_1_2S, dims = 1:125)
DimPlot(cho_1_2S, reduction = "umap")
FeaturePlot(cho_1_2S, features = "WBGene00016114")

umap.cho_1_2 <- cho_1_2S@reductions$umap

cho_1_2_DR <- as.data.frame(Embeddings(umap.cho_1_2))
head(Idents(cho_1_2S))
cho_1_2_DR$Cluster <- Idents(cho_1_2S)
head(cho_1_2_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3441-ST-2/SoupX"
sclcho_1_2 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclcho_1_2$toc[1:5,1:5]
head(rownames(cho_1_2_DR))

rownames(cho_1_2_DR) <- substr(rownames(cho_1_2_DR), 1, 16)
head(cho_1_2_DR)
cho_1_2_DR$aman1 <- sclcho_1_2$toc["WBGene00018877", rownames(cho_1_2_DR)]
gg <- ggplot(cho_1_2_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(sclcho_1_2$toc) <- as.character(vlookup(rownames(sclcho_1_2$toc), gene_list))
sclcho_1_2$toc[1:5,1:5]
rownames(sclcho_1_2$soupProfile) <- rownames(sclcho_1_2$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclcho_1_2, "flp-27", cho_1_2_DR)
plot(gg)
plotMarkerMap(sclcho_1_2, "pdf-1", cho_1_2_DR)
plotMarkerMap(sclcho_1_2, "flp-12", cho_1_2_DR)
plotMarkerMap(sclcho_1_2, "ilys-4", cho_1_2_DR)
plotMarkerMap(sclcho_1_2, "nlp-21", cho_1_2_DR)

plotMarkerMap(sclcho_1_2, geneSet = c("flp-9", "pdf-1", "flp-12"), cho_1_2_DR)
head(sclcho_1_2$soupProfile[order(sclcho_1_2$soupProfile$est, decreasing = TRUE), ], n = 20)

head(cho_1_2_DR)
sclcho_1_2 <- setDR(sclcho_1_2, cho_1_2_DR)

sclcho_1_2 <- setClusters(sclcho_1_2, cho_1_2_DR$Cluster)
plotMarkerMap(sclcho_1_2, "atp-6")
plotMarkerMap(sclcho_1_2, "nlp-64")

plotMarkerDistribution(sclcho_1_2)

cho_1_2.genes <- c("flp-12", "flp-9", "pdf-1", "nlp-64")
useToEst = estimateNonExpressingCells(sclcho_1_2, nonExpressedGeneList = list(cho_1_2 = cho_1_2.genes))
plotMarkerMap(sclcho_1_2, geneSet = cho_1_2.genes, DR = cho_1_2_DR, useToEst = useToEst)
sclcho_1_2 <- calculateContaminationFraction(sclcho_1_2, list(cho_1_2 = cho_1_2.genes), useToEst = useToEst)
# Estimated global contamination fraction of 8.91%

system.time(out.cho_1_2 <- adjustCounts(sclcho_1_2, roundToInt = TRUE))
system.time(out.cho_1_2.ni <- adjustCounts(sclcho_1_2))

cntSoggy = rowSums(sclcho_1_2$toc > 0)
cntStrained = rowSums(out.cho_1_2 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclcho_1_2$toc > out.cho_1_2)/rowSums(sclcho_1_2$toc > 0)), n = 20)

plotChangeMap(sclcho_1_2, out.cho_1_2, "flp-12")
plotChangeMap(sclcho_1_2, out.cho_1_2, "flp-9")
plotChangeMap(sclcho_1_2, out.cho_1_2, "nlp-64")
plotChangeMap(sclcho_1_2, out.cho_1_2, "pdf-1")
plotChangeMap(sclcho_1_2, out.cho_1_2, "ilys-4")
plotChangeMap(sclcho_1_2, out.cho_1_2, "nlp-21")
plotChangeMap(sclcho_1_2, out.cho_1_2, "cho-1")
plotChangeMap(sclcho_1_2, out.cho_1_2, "atp-6")

out.cho_1_2[1:5,1:5]
sclcho_1_2$toc[1:5,1:5]

saveRDS(sclcho_1_2, "./final_gene_model/3441-ST-2/SoupX/soupX_object_cho_1_2.rds")
saveRDS(out.cho_1_2, "./final_gene_model/3441-ST-2/SoupX/soupX_corrected_matrix_cho_1_2.rds")
saveRDS(out.cho_1_2.ni, "./final_gene_model/3441-ST-2/SoupX/soupX_corrected_matrix_cho_1_2_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.cho_1_2) <- paste(colnames(out.cho_1_2), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.cho_1_2) <- as.character(vlookup(rownames(out.cho_1_2), gene_list, result_column = 1, lookup_column = 2))
out.cho_1_2[1:5,1:5]
head(colData(cho_1_2))

cho_1_2_corrected <- SingleCellExperiment(assays = list(counts = out.cho_1_2), colData = colData(cho_1_2))  
cho_1_2
cho_1_2_corrected
head(rowData(cho_1_2))
head(rowData(cho_1_2_corrected))
rowData(cho_1_2_corrected) <- rowData(cho_1_2)
cho_1_2_corrected

cho_1_2 <- calculateQCMetrics(cho_1_2, feature_controls = list(Mito = which(rowData(cho_1_2)$ID == "WBGene00010959" | rowData(cho_1_2)$ID == "WBGene00010961"| rowData(cho_1_2)$ID == "WBGene00010966" | rowData(cho_1_2)$ID == "WBGene00010963" | rowData(cho_1_2)$ID =="WBGene00010957" | rowData(cho_1_2)$ID == "WBGene00010967" | rowData(cho_1_2)$ID == "WBGene00010958" | rowData(cho_1_2)$ID == "WBGene00010960" | rowData(cho_1_2)$ID == "WBGene00010962" | rowData(cho_1_2)$ID == "WBGene00000829" | rowData(cho_1_2)$ID =="WBGene00010965" | rowData(cho_1_2)$ID == "WBGene00010964")))
cho_1_2_corrected <- calculateQCMetrics(cho_1_2_corrected, feature_controls = list(Mito = which(rowData(cho_1_2_corrected)$ID == "WBGene00010959" | rowData(cho_1_2_corrected)$ID == "WBGene00010961"| rowData(cho_1_2_corrected)$ID == "WBGene00010966" | rowData(cho_1_2_corrected)$ID == "WBGene00010963" | rowData(cho_1_2_corrected)$ID =="WBGene00010957" | rowData(cho_1_2_corrected)$ID == "WBGene00010967" | rowData(cho_1_2_corrected)$ID == "WBGene00010958" | rowData(cho_1_2_corrected)$ID == "WBGene00010960" | rowData(cho_1_2_corrected)$ID == "WBGene00010962" | rowData(cho_1_2_corrected)$ID == "WBGene00000829" | rowData(cho_1_2_corrected)$ID =="WBGene00010965" | rowData(cho_1_2_corrected)$ID == "WBGene00010964")))

summary(cho_1_2$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.725   3.054   5.855   6.111  97.380 

summary(cho_1_2_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.0000  0.5385  2.9489  2.1277 97.7217 

table(cho_1_2$pct_counts_Mito < 20)
## FALSE  TRUE
##  253  4501 

table(cho_1_2_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  132  4622  

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(cho_1_2$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(cho_1_2$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(cho_1_2$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(cho_1_2_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(cho_1_2_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(cho_1_2_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(cho_1_2) <- librarySizeFactors(cho_1_2)
cho_1_2 <- normalize(cho_1_2)
cho_1_2
saveRDS(cho_1_2, "./final_gene_model/3441-ST-2/SoupX/020620_cho_1_2_sce_uncorrected.rds")

sizeFactors(cho_1_2_corrected) <- librarySizeFactors(cho_1_2_corrected)
cho_1_2_corrected <- normalize(cho_1_2_corrected)
cho_1_2_corrected
saveRDS(cho_1_2_corrected, "./final_gene_model/3441-ST-2/SoupX/020620_cho_1_2_sce_new_SoupX_corrected.rds")

# 3465-ST-1

my_dir <- "./final_gene_model/3465-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

tph_1 <- read10xCounts(my_dir)
colnames(tph_1) <- tph_1$Barcode
tph_1
head(rowData(tph_1))
rowData(tph_1) <- rowData(tph_1)[,1:2]

rowData(tph_1)$Symbol <- as.character(vlookup(rowData(tph_1)$ID, gene_list))
head(rowData(tph_1))

bcrank <- barcodeRanks(counts(tph_1))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(tph_1), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  28564    5343  703373  

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability", xlim = c(0,1000), ylim = c(0,2000))

c.keep <- defaultDrops(counts(tph_1), expected=5000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##   Both CellRanger EmptyDrops 
## 734808        116       2356 

tph_1$Detection <- detection
tph_1$PValue <- e.out$PValue

tph_1 <- tph_1[, keep]
tph_1
## class: SingleCellExperiment 
## dim: 46911 5459

# Creating a filtered matrix to use for SoupX corrections

tph_1_filtered_counts <- counts(tph_1)
writeMM(tph_1_filtered_counts, "./final_gene_model/3465-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3465-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(tph_1_filtered_counts, "./final_gene_model/3465-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(tph_1)
write.table(fil.barcodes, "./final_gene_model/3465-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3465-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3465-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(tph_1) <- librarySizeFactors(tph_1)
tph_1 <- normalize(tph_1)
tph_1S <- as.Seurat(tph_1)

tph_1S <- NormalizeData(tph_1S)
tph_1S <- FindVariableFeatures(tph_1S, selection.method = "vst", nfeatures = 4000)
tph_1S <- ScaleData(tph_1S)
tph_1S <- RunPCA(tph_1S, features = VariableFeatures(tph_1S), npcs = 145)

ElbowPlot(tph_1S, ndims = 145)

tph_1S <- FindNeighbors(tph_1S, dims = 1:125)
tph_1S <- FindClusters(tph_1S, resolution = 1.2)

tph_1S <- RunUMAP(tph_1S, dims = 1:125)
DimPlot(tph_1S, reduction = "umap")
FeaturePlot(tph_1S, features = "WBGene00000435")

umap.tph_1 <- tph_1S@reductions$umap

tph_1_DR <- as.data.frame(Embeddings(umap.tph_1))
head(Idents(tph_1S))
tph_1_DR$Cluster <- Idents(tph_1S)
head(tph_1_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3465-ST-1/SoupX"
scltph_1 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

scltph_1$toc[1:5,1:5]
head(rownames(tph_1_DR))

rownames(tph_1_DR) <- substr(rownames(tph_1_DR), 1, 16)
head(tph_1_DR)
tph_1_DR$aman1 <- scltph_1$toc["WBGene00018877", rownames(tph_1_DR)]
gg <- ggplot(tph_1_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(scltph_1$toc) <- as.character(vlookup(rownames(scltph_1$toc), gene_list))
scltph_1$toc[1:5,1:5]
rownames(scltph_1$soupProfile) <- rownames(scltph_1$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(scltph_1, "aptf-1", tph_1_DR)
plot(gg)
plotMarkerMap(scltph_1, "R13H7.2", tph_1_DR)
plotMarkerMap(scltph_1, "R102.2", tph_1_DR)
plotMarkerMap(scltph_1, "K10C2.12", tph_1_DR)
plotMarkerMap(scltph_1, "flp-27", tph_1_DR)

plotMarkerMap(scltph_1, geneSet = c("R102.2", "K10C2.12", "acbp-6"), tph_1_DR)
head(scltph_1$soupProfile[order(scltph_1$soupProfile$est, decreasing = TRUE), ], n = 20)

head(tph_1_DR)
scltph_1 <- setDR(scltph_1, tph_1_DR)

scltph_1 <- setClusters(scltph_1, tph_1_DR$Cluster)
plotMarkerMap(scltph_1, "atp-6")
plotMarkerMap(scltph_1, "T05A8.3")


plotMarkerDistribution(scltph_1)

tph_1.genes <- c("R102.2", "T05A8.3", 'flp-14', "pdf-1", "flp-27")
useToEst = estimateNonExpressingCells(scltph_1, nonExpressedGeneList = list(tph_1 = tph_1.genes))
plotMarkerMap(scltph_1, geneSet = tph_1.genes, DR = tph_1_DR, useToEst = useToEst)
scltph_1 <- calculateContaminationFraction(scltph_1, list(tph_1 = tph_1.genes), useToEst = useToEst)
# Estimated global contamination fraction of 2.52%

system.time(out.tph_1 <- adjustCounts(scltph_1, roundToInt = TRUE))
system.time(out.tph_1.ni <- adjustCounts(scltph_1))

cntSoggy = rowSums(scltph_1$toc > 0)
cntStrained = rowSums(out.tph_1 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(scltph_1$toc > out.tph_1)/rowSums(scltph_1$toc > 0)), n = 20)

plotChangeMap(scltph_1, out.tph_1, "acbp-6")
plotChangeMap(scltph_1, out.tph_1, "R102.2")
plotChangeMap(scltph_1, out.tph_1, "T05A8.3")
plotChangeMap(scltph_1, out.tph_1, "pdf-1")
plotChangeMap(scltph_1, out.tph_1, "ilys-4")
plotChangeMap(scltph_1, out.tph_1, "nlp-21")
plotChangeMap(scltph_1, out.tph_1, "cho-1")
plotChangeMap(scltph_1, out.tph_1, "atp-6")

out.tph_1[1:5,1:5]
scltph_1$toc[1:5,1:5]

saveRDS(scltph_1, "./final_gene_model/3465-ST-1/SoupX/soupX_object_tph_1.rds")
saveRDS(out.tph_1, "./final_gene_model/3465-ST-1/SoupX/soupX_corrected_matrix_tph_1.rds")
saveRDS(out.tph_1.ni, "./final_gene_model/3465-ST-1/SoupX/soupX_corrected_matrix_tph_1_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.tph_1) <- paste(colnames(out.tph_1), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.tph_1) <- as.character(vlookup(rownames(out.tph_1), gene_list, result_column = 1, lookup_column = 2))
out.tph_1[1:5,1:5]
head(colData(tph_1))

tph_1_corrected <- SingleCellExperiment(assays = list(counts = out.tph_1), colData = colData(tph_1))  
tph_1
tph_1_corrected
head(rowData(tph_1))
head(rowData(tph_1_corrected))
rowData(tph_1_corrected) <- rowData(tph_1)
tph_1_corrected

tph_1 <- calculateQCMetrics(tph_1, feature_controls = list(Mito = which(rowData(tph_1)$ID == "WBGene00010959" | rowData(tph_1)$ID == "WBGene00010961"| rowData(tph_1)$ID == "WBGene00010966" | rowData(tph_1)$ID == "WBGene00010963" | rowData(tph_1)$ID =="WBGene00010957" | rowData(tph_1)$ID == "WBGene00010967" | rowData(tph_1)$ID == "WBGene00010958" | rowData(tph_1)$ID == "WBGene00010960" | rowData(tph_1)$ID == "WBGene00010962" | rowData(tph_1)$ID == "WBGene00000829" | rowData(tph_1)$ID =="WBGene00010965" | rowData(tph_1)$ID == "WBGene00010964")))
tph_1_corrected <- calculateQCMetrics(tph_1_corrected, feature_controls = list(Mito = which(rowData(tph_1_corrected)$ID == "WBGene00010959" | rowData(tph_1_corrected)$ID == "WBGene00010961"| rowData(tph_1_corrected)$ID == "WBGene00010966" | rowData(tph_1_corrected)$ID == "WBGene00010963" | rowData(tph_1_corrected)$ID =="WBGene00010957" | rowData(tph_1_corrected)$ID == "WBGene00010967" | rowData(tph_1_corrected)$ID == "WBGene00010958" | rowData(tph_1_corrected)$ID == "WBGene00010960" | rowData(tph_1_corrected)$ID == "WBGene00010962" | rowData(tph_1_corrected)$ID == "WBGene00000829" | rowData(tph_1_corrected)$ID =="WBGene00010965" | rowData(tph_1_corrected)$ID == "WBGene00010964")))

summary(tph_1$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.729   2.775   4.285   4.947  54.575

summary(tph_1_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.429   2.437   3.876   4.400  55.638 

table(tph_1$pct_counts_Mito < 20)
## FALSE  TRUE
##  79  5380 

table(tph_1_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  64  5395   

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(tph_1$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(tph_1$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(tph_1$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(tph_1_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(tph_1_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(tph_1_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(tph_1) <- librarySizeFactors(tph_1)
tph_1 <- normalize(tph_1)
tph_1
saveRDS(tph_1, "./final_gene_model/3465-ST-1/SoupX/020620_tph_1_sce_uncorrected.rds")

sizeFactors(tph_1_corrected) <- librarySizeFactors(tph_1_corrected)
tph_1_corrected <- normalize(tph_1_corrected)
tph_1_corrected
saveRDS(tph_1_corrected, "./final_gene_model/3465-ST-1/SoupX/020620_tph_1_sce_new_SoupX_corrected.rds")

# 3495-ST-1

my_dir <- "./final_gene_model/3495-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

acr_2 <- read10xCounts(my_dir)
colnames(acr_2) <- acr_2$Barcode
acr_2
head(rowData(acr_2))
rowData(acr_2) <- rowData(acr_2)[,1:2]

rowData(acr_2)$Symbol <- as.character(vlookup(rowData(acr_2)$ID, gene_list))
head(rowData(acr_2))

bcrank <- barcodeRanks(counts(acr_2))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(acr_2), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  37402   13022 6744456  

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability", xlim = c(0,1000), ylim = c(0,2000))

c.keep <- defaultDrops(counts(acr_2), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##      Both EmptyDrops 
##  6785732       9148 

acr_2$Detection <- detection
acr_2$PValue <- e.out$PValue

acr_2 <- acr_2[, keep]
acr_2
## class: SingleCellExperiment 
## dim: 46911 13022

# Creating a filtered matrix to use for SoupX corrections

acr_2_filtered_counts <- counts(acr_2)
writeMM(acr_2_filtered_counts, "./final_gene_model/3495-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3495-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(acr_2_filtered_counts, "./final_gene_model/3495-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(acr_2)
write.table(fil.barcodes, "./final_gene_model/3495-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3495-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3495-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(acr_2) <- librarySizeFactors(acr_2)
acr_2 <- normalize(acr_2)
acr_2S <- as.Seurat(acr_2)

acr_2S <- NormalizeData(acr_2S)
acr_2S <- FindVariableFeatures(acr_2S, selection.method = "vst", nfeatures = 4000)
acr_2S <- ScaleData(acr_2S)
acr_2S <- RunPCA(acr_2S, features = VariableFeatures(acr_2S), npcs = 145)

ElbowPlot(acr_2S, ndims = 145)

acr_2S <- FindNeighbors(acr_2S, dims = 1:125)
acr_2S <- FindClusters(acr_2S, resolution = 1.2)

acr_2S <- RunUMAP(acr_2S, dims = 1:125)
DimPlot(acr_2S, reduction = "umap")
FeaturePlot(acr_2S, features = "WBGene00000435")

umap.acr_2 <- acr_2S@reductions$umap

acr_2_DR <- as.data.frame(Embeddings(umap.acr_2))
head(Idents(acr_2S))
acr_2_DR$Cluster <- Idents(acr_2S)
head(acr_2_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3495-ST-1/SoupX"
sclacr_2 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclacr_2$toc[1:5,1:5]
head(rownames(acr_2_DR))

rownames(acr_2_DR) <- substr(rownames(acr_2_DR), 1, 16)
head(acr_2_DR)
acr_2_DR$aman1 <- sclacr_2$toc["WBGene00018877", rownames(acr_2_DR)]
gg <- ggplot(acr_2_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(sclacr_2$toc) <- as.character(vlookup(rownames(sclacr_2$toc), gene_list))
sclacr_2$toc[1:5,1:5]
rownames(sclacr_2$soupProfile) <- rownames(sclacr_2$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclacr_2, "nspc-19", acr_2_DR)
plot(gg)
plotMarkerMap(sclacr_2, "nspc-1", acr_2_DR)
plotMarkerMap(sclacr_2, "nspc-14", acr_2_DR)
plotMarkerMap(sclacr_2, "flp-18", acr_2_DR)
plotMarkerMap(sclacr_2, "nspc-20", acr_2_DR)

plotMarkerMap(sclacr_2, geneSet = c("nspc-20", "nspc-14", "nspc-15", "nspc-9", "nspc-19", "nspc-17", "nspc-18", "nspc-10", "nspc-1", "nspc-4", "nspc-5"), acr_2_DR)
head(sclacr_2$soupProfile[order(sclacr_2$soupProfile$est, decreasing = TRUE), ], n = 20)

head(acr_2_DR)
sclacr_2 <- setDR(sclacr_2, acr_2_DR)

sclacr_2 <- setClusters(sclacr_2, acr_2_DR$Cluster)

plotMarkerDistribution(sclacr_2)

acr_2.genes <- c("flp-18", "nspc-20", "nspc-14", "nspc-15", "nspc-9", "nspc-19", "nspc-17", "nspc-18", "nspc-10", "nspc-1", "nspc-4", "nspc-5")
useToEst = estimateNonExpressingCells(sclacr_2, nonExpressedGeneList = list(acr_2 = acr_2.genes))
plotMarkerMap(sclacr_2, geneSet = acr_2.genes, DR = acr_2_DR, useToEst = useToEst)
sclacr_2 <- calculateContaminationFraction(sclacr_2, list(acr_2 = acr_2.genes), useToEst = useToEst)
# Estimated global contamination fraction of 10.00%

system.time(out.acr_2 <- adjustCounts(sclacr_2, roundToInt = TRUE))
system.time(out.acr_2.ni <- adjustCounts(sclacr_2))

cntSoggy = rowSums(sclacr_2$toc > 0)
cntStrained = rowSums(out.acr_2 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclacr_2$toc > out.acr_2)/rowSums(sclacr_2$toc > 0)), n = 20)

plotChangeMap(sclacr_2, out.acr_2, "nspc-20")
plotChangeMap(sclacr_2, out.acr_2, "flp-18")
plotChangeMap(sclacr_2, out.acr_2, "nspc-1")
plotChangeMap(sclacr_2, out.acr_2, "flp-9")
plotChangeMap(sclacr_2, out.acr_2, "ilys-4")
plotChangeMap(sclacr_2, out.acr_2, "nlp-21")
plotChangeMap(sclacr_2, out.acr_2, "C30E1.9")
plotChangeMap(sclacr_2, out.acr_2, "T22E5.1")

out.acr_2[1:5,1:5]
sclacr_2$toc[1:5,1:5]

saveRDS(sclacr_2, "./final_gene_model/3495-ST-1/SoupX/soupX_object_acr_2.rds")
saveRDS(out.acr_2, "./final_gene_model/3495-ST-1/SoupX/soupX_corrected_matrix_acr_2.rds")
saveRDS(out.acr_2.ni, "./final_gene_model/3495-ST-1/SoupX/soupX_corrected_matrix_acr_2_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.acr_2) <- paste(colnames(out.acr_2), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.acr_2) <- as.character(vlookup(rownames(out.acr_2), gene_list, result_column = 1, lookup_column = 2))
out.acr_2[1:5,1:5]
head(colData(acr_2))

acr_2_corrected <- SingleCellExperiment(assays = list(counts = out.acr_2), colData = colData(acr_2))  
acr_2
acr_2_corrected
head(rowData(acr_2))
head(rowData(acr_2_corrected))
rowData(acr_2_corrected) <- rowData(acr_2)
acr_2_corrected

acr_2 <- calculateQCMetrics(acr_2, feature_controls = list(Mito = which(rowData(acr_2)$ID == "WBGene00010959" | rowData(acr_2)$ID == "WBGene00010961"| rowData(acr_2)$ID == "WBGene00010966" | rowData(acr_2)$ID == "WBGene00010963" | rowData(acr_2)$ID =="WBGene00010957" | rowData(acr_2)$ID == "WBGene00010967" | rowData(acr_2)$ID == "WBGene00010958" | rowData(acr_2)$ID == "WBGene00010960" | rowData(acr_2)$ID == "WBGene00010962" | rowData(acr_2)$ID == "WBGene00000829" | rowData(acr_2)$ID =="WBGene00010965" | rowData(acr_2)$ID == "WBGene00010964")))
acr_2_corrected <- calculateQCMetrics(acr_2_corrected, feature_controls = list(Mito = which(rowData(acr_2_corrected)$ID == "WBGene00010959" | rowData(acr_2_corrected)$ID == "WBGene00010961"| rowData(acr_2_corrected)$ID == "WBGene00010966" | rowData(acr_2_corrected)$ID == "WBGene00010963" | rowData(acr_2_corrected)$ID =="WBGene00010957" | rowData(acr_2_corrected)$ID == "WBGene00010967" | rowData(acr_2_corrected)$ID == "WBGene00010958" | rowData(acr_2_corrected)$ID == "WBGene00010960" | rowData(acr_2_corrected)$ID == "WBGene00010962" | rowData(acr_2_corrected)$ID == "WBGene00000829" | rowData(acr_2_corrected)$ID =="WBGene00010965" | rowData(acr_2_corrected)$ID == "WBGene00010964")))

summary(acr_2$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.296   2.161   3.561   3.942  47.026 

summary(acr_2_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.4769  1.2987  2.6519  2.9332 48.4638 

table(acr_2$pct_counts_Mito < 20)
## FALSE  TRUE
##  167 12855 

table(acr_2_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  155 12867  

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(acr_2$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(acr_2$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(acr_2$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(acr_2_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(acr_2_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(acr_2_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(acr_2) <- librarySizeFactors(acr_2)
acr_2 <- normalize(acr_2)
acr_2
saveRDS(acr_2, "./final_gene_model/3495-ST-1/SoupX/020620_acr_2_sce_uncorrected.rds")

sizeFactors(acr_2_corrected) <- librarySizeFactors(acr_2_corrected)
acr_2_corrected <- normalize(acr_2_corrected)
acr_2_corrected
saveRDS(acr_2_corrected, "./final_gene_model/3495-ST-1/SoupX/020620_acr_2_sce_new_SoupX_corrected.rds")

# 3503-ST-1

my_dir <- "./final_gene_model/3503-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

ceh_34 <- read10xCounts(my_dir)
colnames(ceh_34) <- ceh_34$Barcode
ceh_34
head(rowData(ceh_34))
rowData(ceh_34) <- rowData(ceh_34)[,1:2]

rowData(ceh_34)$Symbol <- as.character(vlookup(rowData(ceh_34)$ID, gene_list))
head(rowData(ceh_34))

bcrank <- barcodeRanks(counts(ceh_34))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(ceh_34), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  7588    3555 6783737   

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability", xlim = c(0,1000), ylim = c(0,2000))

c.keep <- defaultDrops(counts(ceh_34), expected=5000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##       Both CellRanger EmptyDrops 
##    6793478          8       1394 

ceh_34$Detection <- detection
ceh_34$PValue <- e.out$PValue

ceh_34 <- ceh_34[, keep]
ceh_34
## class: SingleCellExperiment 
## dim: 46911 3563

# Creating a filtered matrix to use for SoupX corrections

ceh_34_filtered_counts <- counts(ceh_34)
writeMM(ceh_34_filtered_counts, "./final_gene_model/3503-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3503-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(ceh_34_filtered_counts, "./final_gene_model/3503-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(ceh_34)
write.table(fil.barcodes, "./final_gene_model/3503-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3503-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3503-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(ceh_34) <- librarySizeFactors(ceh_34)
ceh_34 <- normalize(ceh_34)
ceh_34S <- as.Seurat(ceh_34)

ceh_34S <- NormalizeData(ceh_34S)
ceh_34S <- FindVariableFeatures(ceh_34S, selection.method = "vst", nfeatures = 4000)
ceh_34S <- ScaleData(ceh_34S)
ceh_34S <- RunPCA(ceh_34S, features = VariableFeatures(ceh_34S), npcs = 125)

ElbowPlot(ceh_34S, ndims = 125)

ceh_34S <- FindNeighbors(ceh_34S, dims = 1:125)
ceh_34S <- FindClusters(ceh_34S, resolution = 1.2)

ceh_34S <- RunUMAP(ceh_34S, dims = 1:125)
DimPlot(ceh_34S, reduction = "umap")
FeaturePlot(ceh_34S, features = "WBGene00000455")

umap.ceh_34 <- ceh_34S@reductions$umap

ceh_34_DR <- as.data.frame(Embeddings(umap.ceh_34))
head(Idents(ceh_34S))
ceh_34_DR$Cluster <- Idents(ceh_34S)
head(ceh_34_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3503-ST-1/SoupX"
sclceh_34 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclceh_34$toc[1:5,1:5]
head(rownames(ceh_34_DR))

rownames(ceh_34_DR) <- substr(rownames(ceh_34_DR), 1, 16)
head(ceh_34_DR)
ceh_34_DR$aman1 <- sclceh_34$toc["WBGene00000455", rownames(ceh_34_DR)]
gg <- ggplot(ceh_34_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(sclceh_34$toc) <- as.character(vlookup(rownames(sclceh_34$toc), gene_list))
sclceh_34$toc[1:5,1:5]
rownames(sclceh_34$soupProfile) <- rownames(sclceh_34$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclceh_34, "trh-1", ceh_34_DR)
plot(gg)
plotMarkerMap(sclceh_34, "cup-4", ceh_34_DR)
plotMarkerMap(sclceh_34, "nlp-3", ceh_34_DR)
plotMarkerMap(sclceh_34, "flp-13", ceh_34_DR)
plotMarkerMap(sclceh_34, "nspc-20", ceh_34_DR)

head(sclceh_34$soupProfile[order(sclceh_34$soupProfile$est, decreasing = TRUE), ], n = 20)

head(ceh_34_DR)
sclceh_34 <- setDR(sclceh_34, ceh_34_DR)

sclceh_34 <- setClusters(sclceh_34, ceh_34_DR$Cluster)
plotMarkerMap(sclceh_34, geneSet = c("nspc-20", "flp-13", "R74.10", "C30E1.9"), ceh_34_DR)
plotMarkerMap(sclceh_34, "R74.10")
plotMarkerMap(sclceh_34, "C30E1.9")
plotMarkerDistribution(sclceh_34)

ceh_34.genes <- c("flp-13", "nspc-20", "nspc-14", "trh-1", "R74.10", "lury-1")
useToEst = estimateNonExpressingCells(sclceh_34, nonExpressedGeneList = list(ceh_34 = ceh_34.genes))
plotMarkerMap(sclceh_34, geneSet = ceh_34.genes, DR = ceh_34_DR, useToEst = useToEst)
sclceh_34 <- calculateContaminationFraction(sclceh_34, list(ceh_34 = ceh_34.genes), useToEst = useToEst)
# Estimated global contamination fraction of 4.69%
system.time(out.ceh_34 <- adjustCounts(sclceh_34, roundToInt = TRUE, nCores = 2))
system.time(out.ceh_34.ni <- adjustCounts(sclceh_34, nCores = 2))

cntSoggy = rowSums(sclceh_34$toc > 0)
cntStrained = rowSums(out.ceh_34 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclceh_34$toc > out.ceh_34)/rowSums(sclceh_34$toc > 0)), n = 20)

plotChangeMap(sclceh_34, out.ceh_34, "nspc-20")
plotChangeMap(sclceh_34, out.ceh_34, "trh-1")
plotChangeMap(sclceh_34, out.ceh_34, "flp-13")
plotChangeMap(sclceh_34, out.ceh_34, "R74.10")

out.ceh_34[1:5,1:5]
sclceh_34$toc[1:5,1:5]

saveRDS(sclceh_34, "./final_gene_model/3503-ST-1/SoupX/soupX_object_ceh_34.rds")
saveRDS(out.ceh_34, "./final_gene_model/3503-ST-1/SoupX/soupX_corrected_matrix_ceh_34.rds")
saveRDS(out.ceh_34.ni, "./final_gene_model/3503-ST-1/SoupX/soupX_corrected_matrix_ceh_34_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.ceh_34) <- paste(colnames(out.ceh_34), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.ceh_34) <- as.character(vlookup(rownames(out.ceh_34), gene_list, result_column = 1, lookup_column = 2))
out.ceh_34[1:5,1:5]
head(colData(ceh_34))

ceh_34_corrected <- SingleCellExperiment(assays = list(counts = out.ceh_34), colData = colData(ceh_34))  
ceh_34
ceh_34_corrected
head(rowData(ceh_34))
head(rowData(ceh_34_corrected))
rowData(ceh_34_corrected) <- rowData(ceh_34)
ceh_34_corrected

ceh_34 <- calculateQCMetrics(ceh_34, feature_controls = list(Mito = which(rowData(ceh_34)$ID == "WBGene00010959" | rowData(ceh_34)$ID == "WBGene00010961"| rowData(ceh_34)$ID == "WBGene00010966" | rowData(ceh_34)$ID == "WBGene00010963" | rowData(ceh_34)$ID =="WBGene00010957" | rowData(ceh_34)$ID == "WBGene00010967" | rowData(ceh_34)$ID == "WBGene00010958" | rowData(ceh_34)$ID == "WBGene00010960" | rowData(ceh_34)$ID == "WBGene00010962" | rowData(ceh_34)$ID == "WBGene00000829" | rowData(ceh_34)$ID =="WBGene00010965" | rowData(ceh_34)$ID == "WBGene00010964")))
ceh_34_corrected <- calculateQCMetrics(ceh_34_corrected, feature_controls = list(Mito = which(rowData(ceh_34_corrected)$ID == "WBGene00010959" | rowData(ceh_34_corrected)$ID == "WBGene00010961"| rowData(ceh_34_corrected)$ID == "WBGene00010966" | rowData(ceh_34_corrected)$ID == "WBGene00010963" | rowData(ceh_34_corrected)$ID =="WBGene00010957" | rowData(ceh_34_corrected)$ID == "WBGene00010967" | rowData(ceh_34_corrected)$ID == "WBGene00010958" | rowData(ceh_34_corrected)$ID == "WBGene00010960" | rowData(ceh_34_corrected)$ID == "WBGene00010962" | rowData(ceh_34_corrected)$ID == "WBGene00000829" | rowData(ceh_34_corrected)$ID =="WBGene00010965" | rowData(ceh_34_corrected)$ID == "WBGene00010964")))

summary(ceh_34$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.057   1.839   4.904   3.907  92.457 

summary(ceh_34_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.00000  0.09787  0.83647  3.72229  2.55911 93.23144

table(ceh_34$pct_counts_Mito < 20)
## FALSE  TRUE
##  244  3319 

table(ceh_34_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  192  3371   

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(ceh_34$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(ceh_34$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(ceh_34$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(ceh_34_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(ceh_34_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(ceh_34_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(ceh_34) <- librarySizeFactors(ceh_34)
ceh_34 <- normalize(ceh_34)
ceh_34
saveRDS(ceh_34, "./final_gene_model/3503-ST-1/SoupX/020620_ceh_34_sce_uncorrected.rds")

sizeFactors(ceh_34_corrected) <- librarySizeFactors(ceh_34_corrected)
ceh_34_corrected <- normalize(ceh_34_corrected)
ceh_34_corrected
saveRDS(ceh_34_corrected, "./final_gene_model/3503-ST-1/SoupX/020620_ceh_34_sce_new_SoupX_corrected.rds")

# 3831-ST-1

my_dir <- "./final_gene_model/3831-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

unc_53 <- read10xCounts(my_dir)
colnames(unc_53) <- unc_53$Barcode
unc_53
head(rowData(unc_53))
rowData(unc_53) <- rowData(unc_53)[,1:2]

rowData(unc_53)$Symbol <- as.character(vlookup(rowData(unc_53)$ID, gene_list))
head(rowData(unc_53))

bcrank <- barcodeRanks(counts(unc_53))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(unc_53), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  22654    5663 6766563  

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability", xlim = c(0,1000), ylim = c(0,2000))

c.keep <- defaultDrops(counts(unc_53), expected=5000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##       Both EmptyDrops 
##  6791212       3668

unc_53$Detection <- detection
unc_53$PValue <- e.out$PValue

unc_53 <- unc_53[, keep]
unc_53
## class: SingleCellExperiment 
## dim: 46911 5663

# Creating a filtered matrix to use for SoupX corrections

unc_53_filtered_counts <- counts(unc_53)
writeMM(unc_53_filtered_counts, "./final_gene_model/3831-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/3831-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(unc_53_filtered_counts, "./final_gene_model/3831-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(unc_53)
write.table(fil.barcodes, "./final_gene_model/3831-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/3831-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/3831-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(unc_53) <- librarySizeFactors(unc_53)
unc_53 <- normalize(unc_53)
unc_53S <- as.Seurat(unc_53)

unc_53S <- NormalizeData(unc_53S)
unc_53S <- FindVariableFeatures(unc_53S, selection.method = "vst", nfeatures = 4000)
unc_53S <- ScaleData(unc_53S)
unc_53S <- RunPCA(unc_53S, features = VariableFeatures(unc_53S), npcs = 145)

ElbowPlot(unc_53S, ndims = 145)

unc_53S <- FindNeighbors(unc_53S, dims = 1:125)
unc_53S <- FindClusters(unc_53S, resolution = 1.2)

unc_53S <- RunUMAP(unc_53S, dims = 1:125)
DimPlot(unc_53S, reduction = "umap")
FeaturePlot(unc_53S, features = "WBGene00000455")

umap.unc_53 <- unc_53S@reductions$umap

unc_53_DR <- as.data.frame(Embeddings(umap.unc_53))
head(Idents(unc_53S))
unc_53_DR$Cluster <- Idents(unc_53S)
head(unc_53_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/3831-ST-1/SoupX"
sclunc_53 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclunc_53$toc[1:5,1:5]
head(rownames(unc_53_DR))

rownames(unc_53_DR) <- substr(rownames(unc_53_DR), 1, 16)
head(unc_53_DR)
unc_53_DR$aman1 <- sclunc_53$toc["WBGene00000455", rownames(unc_53_DR)]
gg <- ggplot(unc_53_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

rownames(sclunc_53$toc) <- as.character(vlookup(rownames(sclunc_53$toc), gene_list))
sclunc_53$toc[1:5,1:5]
rownames(sclunc_53$soupProfile) <- rownames(sclunc_53$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclunc_53, "trh-1", unc_53_DR)
plot(gg)
plotMarkerMap(sclunc_53, "flp-12", unc_53_DR)
plotMarkerMap(sclunc_53, "nlp-64", unc_53_DR)
plotMarkerMap(sclunc_53, "flp-24", unc_53_DR)

head(sclunc_53$soupProfile[order(sclunc_53$soupProfile$est, decreasing = TRUE), ], n = 20)

head(unc_53_DR)
sclunc_53 <- setDR(sclunc_53, unc_53_DR)

sclunc_53 <- setClusters(sclunc_53, unc_53_DR$Cluster)
plotMarkerMap(sclunc_53, geneSet = c("flp-12", "nlp-64", "flp-24", "flp-9", "flp-28"), unc_53_DR)
plotMarkerMap(sclunc_53, "unc-53")
plotMarkerMap(sclunc_53, "C30E1.9")
plotMarkerDistribution(sclunc_53)

unc_53.genes <- c("flp-12", "nlp-64", "flp-24", "flp-9", "flp-28")
useToEst = estimateNonExpressingCells(sclunc_53, nonExpressedGeneList = list(unc_53 = unc_53.genes))
plotMarkerMap(sclunc_53, geneSet = unc_53.genes, DR = unc_53_DR, useToEst = useToEst)
sclunc_53 <- calculateContaminationFraction(sclunc_53, list(unc_53 = unc_53.genes), useToEst = useToEst)
# Estimated global contamination fraction of 5.93%

system.time(out.unc_53 <- adjustCounts(sclunc_53, roundToInt = TRUE, nCores = 2))
system.time(out.unc_53.ni <- adjustCounts(sclunc_53, nCores = 2))

cntSoggy = rowSums(sclunc_53$toc > 0)
cntStrained = rowSums(out.unc_53 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclunc_53$toc > out.unc_53)/rowSums(sclunc_53$toc > 0)), n = 20)

plotChangeMap(sclunc_53, out.unc_53, "flp-9")
plotChangeMap(sclunc_53, out.unc_53, "flp-24")
plotChangeMap(sclunc_53, out.unc_53, "sbt-1")
plotChangeMap(sclunc_53, out.unc_53, "ndk-1")

out.unc_53[1:5,1:5]
sclunc_53$toc[1:5,1:5]

saveRDS(sclunc_53, "./final_gene_model/3831-ST-1/SoupX/soupX_object_unc_53.rds")
saveRDS(out.unc_53, "./final_gene_model/3831-ST-1/SoupX/soupX_corrected_matrix_unc_53.rds")
saveRDS(out.unc_53.ni, "./final_gene_model/3831-ST-1/SoupX/soupX_corrected_matrix_unc_53_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.unc_53) <- paste(colnames(out.unc_53), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.unc_53) <- as.character(vlookup(rownames(out.unc_53), gene_list, result_column = 1, lookup_column = 2))
out.unc_53[1:5,1:5]
head(colData(unc_53))

unc_53_corrected <- SingleCellExperiment(assays = list(counts = out.unc_53), colData = colData(unc_53))  
unc_53
unc_53_corrected
head(rowData(unc_53))
head(rowData(unc_53_corrected))
rowData(unc_53_corrected) <- rowData(unc_53)
unc_53_corrected

unc_53 <- calculateQCMetrics(unc_53, feature_controls = list(Mito = which(rowData(unc_53)$ID == "WBGene00010959" | rowData(unc_53)$ID == "WBGene00010961"| rowData(unc_53)$ID == "WBGene00010966" | rowData(unc_53)$ID == "WBGene00010963" | rowData(unc_53)$ID =="WBGene00010957" | rowData(unc_53)$ID == "WBGene00010967" | rowData(unc_53)$ID == "WBGene00010958" | rowData(unc_53)$ID == "WBGene00010960" | rowData(unc_53)$ID == "WBGene00010962" | rowData(unc_53)$ID == "WBGene00000829" | rowData(unc_53)$ID =="WBGene00010965" | rowData(unc_53)$ID == "WBGene00010964")))
unc_53_corrected <- calculateQCMetrics(unc_53_corrected, feature_controls = list(Mito = which(rowData(unc_53_corrected)$ID == "WBGene00010959" | rowData(unc_53_corrected)$ID == "WBGene00010961"| rowData(unc_53_corrected)$ID == "WBGene00010966" | rowData(unc_53_corrected)$ID == "WBGene00010963" | rowData(unc_53_corrected)$ID =="WBGene00010957" | rowData(unc_53_corrected)$ID == "WBGene00010967" | rowData(unc_53_corrected)$ID == "WBGene00010958" | rowData(unc_53_corrected)$ID == "WBGene00010960" | rowData(unc_53_corrected)$ID == "WBGene00010962" | rowData(unc_53_corrected)$ID == "WBGene00000829" | rowData(unc_53_corrected)$ID =="WBGene00010965" | rowData(unc_53_corrected)$ID == "WBGene00010964")))

summary(unc_53$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.168   2.437   6.131   8.073  73.675 

summary(unc_53_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.9705  2.2140  6.1665  8.1116 77.7533

table(unc_53$pct_counts_Mito < 20)
## FALSE  TRUE
##  392  5271

table(unc_53_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  432  5231  

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(unc_53$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_53$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_53$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(unc_53_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(unc_53_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(unc_53_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(unc_53) <- librarySizeFactors(unc_53)
unc_53 <- normalize(unc_53)
unc_53
saveRDS(unc_53, "./final_gene_model/3831-ST-1/SoupX/020620_unc_53_sce_uncorrected.rds")

sizeFactors(unc_53_corrected) <- librarySizeFactors(unc_53_corrected)
unc_53_corrected <- normalize(unc_53_corrected)
unc_53_corrected
saveRDS(unc_53_corrected, "./final_gene_model/3831-ST-1/SoupX/020620_unc_53_sce_new_SoupX_corrected.rds")


# 4138-ST-1

my_dir <- "./final_gene_model/4138-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

nlp_13 <- read10xCounts(my_dir)
colnames(nlp_13) <- nlp_13$Barcode
nlp_13
head(rowData(nlp_13))
rowData(nlp_13) <- rowData(nlp_13)[,1:2]

rowData(nlp_13)$Symbol <- as.character(vlookup(rowData(nlp_13)$ID, gene_list))
head(rowData(nlp_13))

bcrank <- barcodeRanks(counts(nlp_13))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(nlp_13), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical  25199    9858 6759823 

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability", xlim = c(0,1000), ylim = c(0,2000))

c.keep <- defaultDrops(counts(nlp_13), expected=5000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##       Both EmptyDrops 
##    6788348       6532 

nlp_13$Detection <- detection
nlp_13$PValue <- e.out$PValue

nlp_13 <- nlp_13[, keep]
nlp_13
## class: SingleCellExperiment 
## dim: 46911 9858

# Creating a filtered matrix to use for SoupX corrections

nlp_13_filtered_counts <- counts(nlp_13)
writeMM(nlp_13_filtered_counts, "./final_gene_model/4138-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/4138-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(nlp_13_filtered_counts, "./final_gene_model/4138-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(nlp_13)
write.table(fil.barcodes, "./final_gene_model/4138-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/4138-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/4138-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(nlp_13) <- librarySizeFactors(nlp_13)
nlp_13 <- normalize(nlp_13)
nlp_13S <- as.Seurat(nlp_13)

nlp_13S <- NormalizeData(nlp_13S)
nlp_13S <- FindVariableFeatures(nlp_13S, selection.method = "vst", nfeatures = 4000)
nlp_13S <- ScaleData(nlp_13S)
nlp_13S <- RunPCA(nlp_13S, features = VariableFeatures(nlp_13S), npcs = 145)

ElbowPlot(nlp_13S, ndims = 145)

nlp_13S <- FindNeighbors(nlp_13S, dims = 1:125)
nlp_13S <- FindClusters(nlp_13S, resolution = 1.2)

nlp_13S <- RunUMAP(nlp_13S, dims = 1:125)
DimPlot(nlp_13S, reduction = "umap")
FeaturePlot(nlp_13S, features = "WBGene00001451")

umap.nlp_13 <- nlp_13S@reductions$umap

nlp_13_DR <- as.data.frame(Embeddings(umap.nlp_13))
head(Idents(nlp_13S))
nlp_13_DR$Cluster <- Idents(nlp_13S)
head(nlp_13_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/4138-ST-1/SoupX"
sclnlp_13 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclnlp_13$toc[1:5,1:5]
head(rownames(nlp_13_DR))

rownames(nlp_13_DR) <- substr(rownames(nlp_13_DR), 1, 16)
head(nlp_13_DR)
nlp_13_DR$aman1 <- sclnlp_13$toc["WBGene00000455", rownames(nlp_13_DR)]
gg <- ggplot(nlp_13_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

gene_list <- read.table("./final_gene_model/genes.txt", sep = "\t")

rownames(sclnlp_13$toc) <- as.character(vlookup(rownames(sclnlp_13$toc), gene_list))

sclnlp_13$toc[1:5,1:5]
rownames(sclnlp_13$soupProfile) <- rownames(sclnlp_13$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclnlp_13, "flp-8", nlp_13_DR)
plot(gg)
plotMarkerMap(sclnlp_13, "flp-24", nlp_13_DR)
plotMarkerMap(sclnlp_13, "col-93", nlp_13_DR)
plotMarkerMap(sclnlp_13, "", nlp_13_DR)

head(sclnlp_13$soupProfile[order(sclnlp_13$soupProfile$est, decreasing = TRUE), ], n = 20)
plotMarkerMap(sclnlp_13, "D1054.10", nlp_13_DR)
plotMarkerMap(sclnlp_13, "Y37D8A.19", nlp_13_DR)
plotMarkerMap(sclnlp_13, "ttr-14", nlp_13_DR)
plotMarkerMap(sclnlp_13, "nlp-18", nlp_13_DR)

head(nlp_13_DR)
sclnlp_13 <- setDR(sclnlp_13, nlp_13_DR)

sclnlp_13 <- setClusters(sclnlp_13, nlp_13_DR$Cluster)
plotMarkerMap(sclnlp_13, geneSet = c("D1054.10", "Y37D8A.19", "ttr-14", "nlp-18"), nlp_13_DR)
plotMarkerMap(sclnlp_13, "Y62H9A.5")
plotMarkerMap(sclnlp_13, "ule-3")
plotMarkerMap(sclnlp_13, "ZC373.2")

plotMarkerMap(sclnlp_13, "R102.2")
plotMarkerDistribution(sclnlp_13)

nlp_13.genes <- c("D1054.10", "Y37D8A.19", "ttr-14", "Y62H9A.5","ule-3", "ZC373.2", "nlp-18", "R102.2")
useToEst = estimateNonExpressingCells(sclnlp_13, nonExpressedGeneList = list(nlp_13 = nlp_13.genes))
plotMarkerMap(sclnlp_13, geneSet = nlp_13.genes, DR = nlp_13_DR, useToEst = useToEst)
sclnlp_13 <- calculateContaminationFraction(sclnlp_13, list(nlp_13 = nlp_13.genes), useToEst = useToEst)
# Estimated global contamination fraction of 4.15%
system.time(out.nlp_13 <- adjustCounts(sclnlp_13, roundToInt = TRUE))
system.time(out.nlp_13.ni <- adjustCounts(sclnlp_13))

cntSoggy = rowSums(sclnlp_13$toc > 0)
cntStrained = rowSums(out.nlp_13 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclnlp_13$toc > out.nlp_13)/rowSums(sclnlp_13$toc > 0)), n = 20)

plotChangeMap(sclnlp_13, out.nlp_13, "flp-8")
plotChangeMap(sclnlp_13, out.nlp_13, "R102.2")
plotChangeMap(sclnlp_13, out.nlp_13, "ule-3")
plotChangeMap(sclnlp_13, out.nlp_13, "ttr-14")

out.nlp_13[1:5,1:5]
sclnlp_13$toc[1:5,1:5]

saveRDS(sclnlp_13, "./final_gene_model/4138-ST-1/SoupX/soupX_object_nlp_13.rds")
saveRDS(out.nlp_13, "./final_gene_model/4138-ST-1/SoupX/soupX_corrected_matrix_nlp_13.rds")
saveRDS(out.nlp_13.ni, "./final_gene_model/4138-ST-1/SoupX/soupX_corrected_matrix_nlp_13_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.nlp_13) <- paste(colnames(out.nlp_13), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.nlp_13) <- as.character(vlookup(rownames(out.nlp_13), gene_list, result_column = 1, lookup_column = 2))
out.nlp_13[1:5,1:5]
head(colData(nlp_13))

nlp_13_corrected <- SingleCellExperiment(assays = list(counts = out.nlp_13), colData = colData(nlp_13))  
nlp_13
nlp_13_corrected
head(rowData(nlp_13))
head(rowData(nlp_13_corrected))
rowData(nlp_13_corrected) <- rowData(nlp_13)
nlp_13_corrected

nlp_13 <- calculateQCMetrics(nlp_13, feature_controls = list(Mito = which(rowData(nlp_13)$ID == "WBGene00010959" | rowData(nlp_13)$ID == "WBGene00010961"| rowData(nlp_13)$ID == "WBGene00010966" | rowData(nlp_13)$ID == "WBGene00010963" | rowData(nlp_13)$ID =="WBGene00010957" | rowData(nlp_13)$ID == "WBGene00010967" | rowData(nlp_13)$ID == "WBGene00010958" | rowData(nlp_13)$ID == "WBGene00010960" | rowData(nlp_13)$ID == "WBGene00010962" | rowData(nlp_13)$ID == "WBGene00000829" | rowData(nlp_13)$ID =="WBGene00010965" | rowData(nlp_13)$ID == "WBGene00010964")))
nlp_13_corrected <- calculateQCMetrics(nlp_13_corrected, feature_controls = list(Mito = which(rowData(nlp_13_corrected)$ID == "WBGene00010959" | rowData(nlp_13_corrected)$ID == "WBGene00010961"| rowData(nlp_13_corrected)$ID == "WBGene00010966" | rowData(nlp_13_corrected)$ID == "WBGene00010963" | rowData(nlp_13_corrected)$ID =="WBGene00010957" | rowData(nlp_13_corrected)$ID == "WBGene00010967" | rowData(nlp_13_corrected)$ID == "WBGene00010958" | rowData(nlp_13_corrected)$ID == "WBGene00010960" | rowData(nlp_13_corrected)$ID == "WBGene00010962" | rowData(nlp_13_corrected)$ID == "WBGene00000829" | rowData(nlp_13_corrected)$ID =="WBGene00010965" | rowData(nlp_13_corrected)$ID == "WBGene00010964")))

summary(nlp_13$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.040   2.138   4.799   5.556  50.785  

summary(nlp_13_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.4665  1.5205  4.1611  4.6462 51.3369

table(nlp_13$pct_counts_Mito < 20)
## FALSE  TRUE
##  467  9391 

table(nlp_13_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  440  9418   

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(nlp_13$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(nlp_13$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(nlp_13$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(nlp_13_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(nlp_13_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(nlp_13_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(nlp_13) <- librarySizeFactors(nlp_13)
nlp_13 <- normalize(nlp_13)
nlp_13
saveRDS(nlp_13, "./final_gene_model/4138-ST-1/SoupX/020620_nlp_13_sce_uncorrected.rds")

sizeFactors(nlp_13_corrected) <- librarySizeFactors(nlp_13_corrected)
nlp_13_corrected <- normalize(nlp_13_corrected)
nlp_13_corrected
saveRDS(nlp_13_corrected, "./final_gene_model/4138-ST-1/SoupX/020620_nlp_13_sce_new_SoupX_corrected.rds")



# 4170-ST-1

my_dir <- "./final_gene_model/4170-ST-1/SoupX/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

ceh_28 <- read10xCounts(my_dir)
colnames(ceh_28) <- ceh_28$Barcode
ceh_28
head(rowData(ceh_28))
rowData(ceh_28) <- rowData(ceh_28)[,1:2]

rowData(ceh_28)$Symbol <- as.character(vlookup(rowData(ceh_28)$ID, gene_list))
head(rowData(ceh_28))

bcrank <- barcodeRanks(counts(ceh_28))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(ceh_28), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical 61250    7920 6725710 

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability", xlim = c(0,1000), ylim = c(0,2000))
remove
c.keep <- defaultDrops(counts(ceh_28), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##       Both EmptyDrops 
##    6788348       6532 

ceh_28$Detection <- detection
ceh_28$PValue <- e.out$PValue

ceh_28 <- ceh_28[, keep]
ceh_28
## class: SingleCellExperiment 
## dim: 46911 7944

# Creating a filtered matrix to use for SoupX corrections

ceh_28_filtered_counts <- counts(ceh_28)
writeMM(ceh_28_filtered_counts, "./final_gene_model/4170-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/4170-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
writeMM(ceh_28_filtered_counts, "./final_gene_model/4170-ST-1/SoupX/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(ceh_28)
write.table(fil.barcodes, "./final_gene_model/4170-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/4170-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/4170-ST-1/SoupX/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(ceh_28) <- librarySizeFactors(ceh_28)
ceh_28 <- normalize(ceh_28)
ceh_28S <- as.Seurat(ceh_28)

ceh_28S <- NormalizeData(ceh_28S)
ceh_28S <- FindVariableFeatures(ceh_28S, selection.method = "vst", nfeatures = 4000)
ceh_28S <- ScaleData(ceh_28S)
ceh_28S <- RunPCA(ceh_28S, features = VariableFeatures(ceh_28S), npcs = 145)

ElbowPlot(ceh_28S, ndims = 145)

ceh_28S <- FindNeighbors(ceh_28S, dims = 1:125)
ceh_28S <- FindClusters(ceh_28S, resolution = 1.2)

ceh_28S <- RunUMAP(ceh_28S, dims = 1:125)
DimPlot(ceh_28S, reduction = "umap")
FeaturePlot(ceh_28S, features = "WBGene00001451")

umap.ceh_28 <- ceh_28S@reductions$umap

ceh_28_DR <- as.data.frame(Embeddings(umap.ceh_28))
head(Idents(ceh_28S))
ceh_28_DR$Cluster <- Idents(ceh_28S)
head(ceh_28_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/4170-ST-1/SoupX"
sclceh_28 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclceh_28$toc[1:5,1:5]
head(rownames(ceh_28_DR))

rownames(ceh_28_DR) <- substr(rownames(ceh_28_DR), 1, 16)
head(ceh_28_DR)
ceh_28_DR$aman1 <- sclceh_28$toc["WBGene00000455", rownames(ceh_28_DR)]
gg <- ggplot(ceh_28_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = aman1 > 0))
plot(gg)

gene_list <- read.table("./final_gene_model/genes.txt", sep = "\t")

rownames(sclceh_28$toc) <- as.character(vlookup(rownames(sclceh_28$toc), gene_list))

sclceh_28$toc[1:5,1:5]
rownames(sclceh_28$soupProfile) <- rownames(sclceh_28$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclceh_28, "mec-17", ceh_28_DR)
plot(gg)
plotMarkerMap(sclceh_28, "mec-3", ceh_28_DR)
plotMarkerMap(sclceh_28, "nlp-18", ceh_28_DR)
plotMarkerMap(sclceh_28, "R102.2", ceh_28_DR)

head(sclceh_28$soupProfile[order(sclceh_28$soupProfile$est, decreasing = TRUE), ], n = 20)
plotMarkerMap(sclceh_28, "pat-10", ceh_28_DR)
plotMarkerMap(sclceh_28, "lev-11", ceh_28_DR)

head(ceh_28_DR)
sclceh_28 <- setDR(sclceh_28, ceh_28_DR)

sclceh_28 <- setClusters(sclceh_28, ceh_28_DR$Cluster)
plotMarkerMap(sclceh_28, geneSet = c("lev-11", "pat-10", "cpn-3", "mec-17", "mec-3", "nlp-18"), ceh_28_DR)

plotMarkerMap(sclceh_28, "R102.2")
plotMarkerDistribution(sclceh_28)

ceh_28.genes <- c("lev-11", "pat-10", "cpn-3", "mec-17", "mec-3", "nlp-18", "R102.2", "mec-10")
useToEst = estimateNonExpressingCells(sclceh_28, nonExpressedGeneList = list(ceh_28 = ceh_28.genes))
plotMarkerMap(sclceh_28, geneSet = ceh_28.genes, DR = ceh_28_DR, useToEst = useToEst)
sclceh_28 <- calculateContaminationFraction(sclceh_28, list(ceh_28 = ceh_28.genes), useToEst = useToEst)
# Estimated global contamination fraction of 4.82%
system.time(out.ceh_28 <- adjustCounts(sclceh_28, roundToInt = TRUE))
system.time(out.ceh_28.ni <- adjustCounts(sclceh_28))

cntSoggy = rowSums(sclceh_28$toc > 0)
cntStrained = rowSums(out.ceh_28 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclceh_28$toc > out.ceh_28)/rowSums(sclceh_28$toc > 0)), n = 20)

plotChangeMap(sclceh_28, out.ceh_28, "pat-10")
plotChangeMap(sclceh_28, out.ceh_28, "R102.2")
plotChangeMap(sclceh_28, out.ceh_28, "mec-17")
plotChangeMap(sclceh_28, out.ceh_28, "cpn-3")

out.ceh_28[1:5,1:5]
sclceh_28$toc[1:5,1:5]

saveRDS(sclceh_28, "./final_gene_model/4170-ST-1/SoupX/soupX_object_ceh_28.rds")
saveRDS(out.ceh_28, "./final_gene_model/4170-ST-1/SoupX/soupX_corrected_matrix_ceh_28.rds")
saveRDS(out.ceh_28.ni, "./final_gene_model/4170-ST-1/SoupX/soupX_corrected_matrix_ceh_28_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.ceh_28) <- paste(colnames(out.ceh_28), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.ceh_28) <- as.character(vlookup(rownames(out.ceh_28), gene_list, result_column = 1, lookup_column = 2))
out.ceh_28[1:5,1:5]
head(colData(ceh_28))

ceh_28_corrected <- SingleCellExperiment(assays = list(counts = out.ceh_28), colData = colData(ceh_28))  
ceh_28
ceh_28_corrected
head(rowData(ceh_28))
head(rowData(ceh_28_corrected))
rowData(ceh_28_corrected) <- rowData(ceh_28)
ceh_28_corrected

ceh_28 <- calculateQCMetrics(ceh_28, feature_controls = list(Mito = which(rowData(ceh_28)$ID == "WBGene00010959" | rowData(ceh_28)$ID == "WBGene00010961"| rowData(ceh_28)$ID == "WBGene00010966" | rowData(ceh_28)$ID == "WBGene00010963" | rowData(ceh_28)$ID =="WBGene00010957" | rowData(ceh_28)$ID == "WBGene00010967" | rowData(ceh_28)$ID == "WBGene00010958" | rowData(ceh_28)$ID == "WBGene00010960" | rowData(ceh_28)$ID == "WBGene00010962" | rowData(ceh_28)$ID == "WBGene00000829" | rowData(ceh_28)$ID =="WBGene00010965" | rowData(ceh_28)$ID == "WBGene00010964")))
ceh_28_corrected <- calculateQCMetrics(ceh_28_corrected, feature_controls = list(Mito = which(rowData(ceh_28_corrected)$ID == "WBGene00010959" | rowData(ceh_28_corrected)$ID == "WBGene00010961"| rowData(ceh_28_corrected)$ID == "WBGene00010966" | rowData(ceh_28_corrected)$ID == "WBGene00010963" | rowData(ceh_28_corrected)$ID =="WBGene00010957" | rowData(ceh_28_corrected)$ID == "WBGene00010967" | rowData(ceh_28_corrected)$ID == "WBGene00010958" | rowData(ceh_28_corrected)$ID == "WBGene00010960" | rowData(ceh_28_corrected)$ID == "WBGene00010962" | rowData(ceh_28_corrected)$ID == "WBGene00000829" | rowData(ceh_28_corrected)$ID =="WBGene00010965" | rowData(ceh_28_corrected)$ID == "WBGene00010964")))

summary(ceh_28$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.635   3.475   5.910   8.271  51.088  

summary(ceh_28_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.066   2.784   5.356   7.493  53.054 

table(ceh_28$pct_counts_Mito < 20)
## FALSE  TRUE
##  324  7620 

table(ceh_28_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  330  7614  

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(ceh_28$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(ceh_28$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(ceh_28$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(ceh_28_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(ceh_28_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(ceh_28_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(ceh_28) <- librarySizeFactors(ceh_28)
ceh_28 <- normalize(ceh_28)
ceh_28
saveRDS(ceh_28, "./final_gene_model/4170-ST-1/SoupX/020620_ceh_28_sce_uncorrected.rds")

sizeFactors(ceh_28_corrected) <- librarySizeFactors(ceh_28_corrected)
ceh_28_corrected <- normalize(ceh_28_corrected)
ceh_28_corrected
saveRDS(ceh_28_corrected, "./final_gene_model/4170-ST-1/SoupX/020620_ceh_28_sce_new_SoupX_corrected.rds")

# 2/13/20 - I noticed that this experiment possibly had more contamination than others, so I wanted to re-try it,
# from a new, clean R session

library(DropletUtils)
library(scater)
library(dplyr)
library(expss)

setwd("~/Dropbox (VU Basic Sciences)/Miller lab/10X Genomics")
# 4170-ST-1

my_dir <- "./final_gene_model/4170-ST-1/SoupX2/raw_feature_bc_matrix"

# I copied the barcode and gene tsv files into a new directory called filtered_feature_bc_matrix. 
# Now I will call EmptyDrops on the raw matrix, generate the filtered matrix, then extract it and save it.

ceh_28 <- read10xCounts(my_dir)
colnames(ceh_28) <- ceh_28$Barcode
ceh_28
head(rowData(ceh_28))
rowData(ceh_28) <- rowData(ceh_28)[,1:2]

rowData(ceh_28)$Symbol <- as.character(vlookup(rowData(ceh_28)$ID, gene_list))
head(rowData(ceh_28))

bcrank <- barcodeRanks(counts(ceh_28))
uniq <- !duplicated(bcrank$rank)

plot(bcrank$rank[uniq], bcrank$total[uniq], log = "xy", xlab = "Rank", ylab = "Total UMI count", cex_lab = 1.2)

set.seed(100)
e.out <- emptyDrops(counts(ceh_28), lower = 50)
summary(e.out$FDR<0.01)
# result from lower - 50
# Mode    FALSE    TRUE    NA's 
# logical 61250    7920 6725710 

is.cell <- e.out$FDR <= 0.01
sum(is.cell, na.rm=TRUE)
table(Limited=e.out$Limited, Significant=is.cell)
par(mfrow=c(1,1))
plot(e.out$Total, -e.out$LogProb, col=ifelse(is.cell, "red", "black"),
     xlab="Total UMI count", ylab="-Log Probability", xlim = c(0,1000), ylim = c(0,2000))

c.keep <- defaultDrops(counts(ceh_28), expected=10000)
e.keep <- e.out$FDR <=0.01
keep <- c.keep | (e.keep & !is.na(e.keep))
detection <- rep("Both", length(keep))
detection[c.keep & !e.keep] <- "CellRanger"
detection[!c.keep & e.keep] <- "EmptyDrops"
table(detection)
##      Both CellRanger EmptyDrops 
## 6790185         54       4641 

ceh_28$Detection <- detection
ceh_28$PValue <- e.out$PValue

ceh_28 <- ceh_28[, keep]
ceh_28
## class: SingleCellExperiment 
## dim: 46911 7974

# Creating a filtered matrix to use for SoupX corrections

ceh_28_filtered_counts <- counts(ceh_28)
writeMM(ceh_28_filtered_counts, "./final_gene_model/4170-ST-1/SoupX2/filtered_feature_bc_matrix/matrix.mtx")
system("gzip ./final_gene_model/4170-ST-1/SoupX2/filtered_feature_bc_matrix/matrix.mtx")
writeMM(ceh_28_filtered_counts, "./final_gene_model/4170-ST-1/SoupX2/filtered_feature_bc_matrix/matrix.mtx")
fil.barcodes <- colnames(ceh_28)
write.table(fil.barcodes, "./final_gene_model/4170-ST-1/SoupX2/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
system("gzip ./final_gene_model/4170-ST-1/SoupX2/filtered_feature_bc_matrix/barcodes.tsv")
write.table(fil.barcodes, "./final_gene_model/4170-ST-1/SoupX2/filtered_feature_bc_matrix/barcodes.tsv", sep = "\t", quote = FALSE, col.names =FALSE, row.names = FALSE)

sizeFactors(ceh_28) <- librarySizeFactors(ceh_28)
ceh_28 <- normalize(ceh_28)
ceh_28S <- as.Seurat(ceh_28)

ceh_28S <- NormalizeData(ceh_28S)
ceh_28S <- FindVariableFeatures(ceh_28S, selection.method = "vst", nfeatures = 4000)
ceh_28S <- ScaleData(ceh_28S)
ceh_28S <- RunPCA(ceh_28S, features = VariableFeatures(ceh_28S), npcs = 145)

ElbowPlot(ceh_28S, ndims = 145)

ceh_28S <- FindNeighbors(ceh_28S, dims = 1:125)
ceh_28S <- FindClusters(ceh_28S, resolution = 1.2)

ceh_28S <- RunUMAP(ceh_28S, dims = 1:125)
DimPlot(ceh_28S, reduction = "umap")
FeaturePlot(ceh_28S, features = "WBGene00001451")

umap.ceh_28 <- ceh_28S@reductions$umap

ceh_28_DR <- as.data.frame(Embeddings(umap.ceh_28))
head(Idents(ceh_28S))
ceh_28_DR$Cluster <- Idents(ceh_28S)
head(ceh_28_DR)

# Now I will try running SoupX, with soupRange = c(0, 25)

my_dir_S <- "./final_gene_model/4170-ST-1/SoupX2"
sclceh_28 <- load10X(my_dir_S, soupRange = c(0, 25))

# This time I used the features.tsv file from the CellRanger output, but that doesn't have the gene short names
# It needs a zipped file, so I could just use the version I created a zip a copy of it, then that should work, and maybe keep the gene names.

# New version is benefitted by clustering

# Newer method

sclceh_28$toc[1:5,1:5]
head(rownames(ceh_28_DR))

rownames(ceh_28_DR) <- substr(rownames(ceh_28_DR), 1, 16)
head(ceh_28_DR)
ceh_28_DR$mec17 <- sclceh_28$toc["WBGene00003178", rownames(ceh_28_DR)]
gg <- ggplot(ceh_28_DR, aes(UMAP_1, UMAP_2)) + geom_point(aes(colour = mec17 > 0))
plot(gg)

gene_list <- read.table("./final_gene_model/genes.txt", sep = "\t")

rownames(sclceh_28$toc) <- as.character(vlookup(rownames(sclceh_28$toc), gene_list))

sclceh_28$toc[1:5,1:5]
rownames(sclceh_28$soupProfile) <- rownames(sclceh_28$toc)
# I changed the rownames of the soupProfile to match rownames of the toc file. They were different. One was gene name, the other id
# Let's see if it changes anything.

gg = plotMarkerMap(sclceh_28, "mec-17", ceh_28_DR)
plot(gg)
plotMarkerMap(sclceh_28, "mec-3", ceh_28_DR)
plotMarkerMap(sclceh_28, "mec-17", ceh_28_DR)
plotMarkerMap(sclceh_28, "mec-7", ceh_28_DR)
plotMarkerMap(sclceh_28, "mec-10", ceh_28_DR)
plotMarkerMap(sclceh_28, "mec-12", ceh_28_DR)

head(sclceh_28$soupProfile[order(sclceh_28$soupProfile$est, decreasing = TRUE), ], n = 20)
plotMarkerMap(sclceh_28, "pat-10", ceh_28_DR)
plotMarkerMap(sclceh_28, "lev-11", ceh_28_DR)
plotMarkerMap(sclceh_28, "R102.2", ceh_28_DR)

head(ceh_28_DR)
sclceh_28 <- setDR(sclceh_28, ceh_28_DR)

sclceh_28 <- setClusters(sclceh_28, ceh_28_DR$Cluster)
plotMarkerMap(sclceh_28, geneSet = c("pat-10", "mec-17", "mec-3", "mec-12", "mec-10"), ceh_28_DR)

plotMarkerDistribution(sclceh_28)

ceh_28.genes <- c("pat-10", "mec-17", "mec-7", "mlc-3", "mec-10", "lev-11", "cpn-3", "nlp-18", "R102.2", "F09E10.7")
useToEst = estimateNonExpressingCells(sclceh_28, nonExpressedGeneList = list(ceh_28 = ceh_28.genes))
plotMarkerMap(sclceh_28, geneSet = ceh_28.genes, DR = ceh_28_DR, useToEst = useToEst)
sclceh_28 <- calculateContaminationFraction(sclceh_28, list(ceh_28 = ceh_28.genes), useToEst = useToEst)
# Estimated global contamination fraction of 4.96%

# Trying different genes didn't change much, but the set of cells being used to estimate is small.
# I am going to try setting a higher estimation, of 10%

sclceh_28 <- setContaminationFraction(sclceh_28, 0.1)

system.time(out.ceh_28 <- adjustCounts(sclceh_28, roundToInt = TRUE))
system.time(out.ceh_28.ni <- adjustCounts(sclceh_28))

cntSoggy = rowSums(sclceh_28$toc > 0)
cntStrained = rowSums(out.ceh_28 > 0)
mostZeroed = tail(sort((cntSoggy - cntStrained)/cntSoggy), n = 10)
mostZeroed

tail(sort(rowSums(sclceh_28$toc > out.ceh_28)/rowSums(sclceh_28$toc > 0)), n = 20)

plotChangeMap(sclceh_28, out.ceh_28, "pat-10")
plotChangeMap(sclceh_28, out.ceh_28, "R102.2")
plotChangeMap(sclceh_28, out.ceh_28, "mec-17")
plotChangeMap(sclceh_28, out.ceh_28, "cpn-3")

out.ceh_28[1:5,1:5]
sclceh_28$toc[1:5,1:5]

saveRDS(sclceh_28, "./final_gene_model/4170-ST-1/SoupX2/soupX_object_ceh_28.rds")
saveRDS(out.ceh_28, "./final_gene_model/4170-ST-1/SoupX2/soupX_corrected_matrix_ceh_28.rds")
saveRDS(out.ceh_28.ni, "./final_gene_model/4170-ST-1/SoupX2/soupX_corrected_matrix_ceh_28_non_integer.rds")

# out object is the corrected matrix, but the "-1" suffix on the barcodes was dropped.  I can add it back in to make it compatible with the meta.data

colnames(out.ceh_28) <- paste(colnames(out.ceh_28), '1', sep = "-")

# I need to make the rownames consistent also
rownames(out.ceh_28) <- as.character(vlookup(rownames(out.ceh_28), gene_list, result_column = 1, lookup_column = 2))
out.ceh_28[1:5,1:5]
head(colData(ceh_28))

ceh_28_corrected <- SingleCellExperiment(assays = list(counts = out.ceh_28), colData = colData(ceh_28))  
ceh_28
ceh_28_corrected
head(rowData(ceh_28))
head(rowData(ceh_28_corrected))
rowData(ceh_28_corrected) <- rowData(ceh_28)
ceh_28_corrected

ceh_28 <- calculateQCMetrics(ceh_28, feature_controls = list(Mito = which(rowData(ceh_28)$ID == "WBGene00010959" | rowData(ceh_28)$ID == "WBGene00010961"| rowData(ceh_28)$ID == "WBGene00010966" | rowData(ceh_28)$ID == "WBGene00010963" | rowData(ceh_28)$ID =="WBGene00010957" | rowData(ceh_28)$ID == "WBGene00010967" | rowData(ceh_28)$ID == "WBGene00010958" | rowData(ceh_28)$ID == "WBGene00010960" | rowData(ceh_28)$ID == "WBGene00010962" | rowData(ceh_28)$ID == "WBGene00000829" | rowData(ceh_28)$ID =="WBGene00010965" | rowData(ceh_28)$ID == "WBGene00010964")))
ceh_28_corrected <- calculateQCMetrics(ceh_28_corrected, feature_controls = list(Mito = which(rowData(ceh_28_corrected)$ID == "WBGene00010959" | rowData(ceh_28_corrected)$ID == "WBGene00010961"| rowData(ceh_28_corrected)$ID == "WBGene00010966" | rowData(ceh_28_corrected)$ID == "WBGene00010963" | rowData(ceh_28_corrected)$ID =="WBGene00010957" | rowData(ceh_28_corrected)$ID == "WBGene00010967" | rowData(ceh_28_corrected)$ID == "WBGene00010958" | rowData(ceh_28_corrected)$ID == "WBGene00010960" | rowData(ceh_28_corrected)$ID == "WBGene00010962" | rowData(ceh_28_corrected)$ID == "WBGene00000829" | rowData(ceh_28_corrected)$ID =="WBGene00010965" | rowData(ceh_28_corrected)$ID == "WBGene00010964")))

summary(ceh_28$pct_counts_Mito)
##  Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.000   1.634   3.473   5.916   8.271  91.029 

summary(ceh_28_corrected$pct_counts_Mito)
## Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
## 0.0000  0.3377  1.6672  4.3970  5.9663 91.6111 

table(ceh_28$pct_counts_Mito < 20)
## FALSE  TRUE
##   325  7649 

table(ceh_28_corrected$pct_counts_Mito < 20)
## FALSE TRUE
##  291  7683  

#Visualizing general count distribution
par(mfrow=c(1,3))
hist(ceh_28$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(ceh_28$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(ceh_28$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

hist(ceh_28_corrected$log10_total_counts, breaks=20, col="grey80",
     xlab="Log-total UMI count")
hist(ceh_28_corrected$log10_total_features_by_counts, breaks=20, col="grey80",
     xlab="Log-total number of expressed features")
hist(ceh_28_corrected$pct_counts_Mito, breaks=20, col="grey80",
     xlab="Proportion of reads in mitochondrial genes")

sizeFactors(ceh_28) <- librarySizeFactors(ceh_28)
ceh_28 <- normalize(ceh_28)
ceh_28
saveRDS(ceh_28, "./final_gene_model/4170-ST-1/SoupX2/020620_ceh_28_sce_uncorrected.rds")

sizeFactors(ceh_28_corrected) <- librarySizeFactors(ceh_28_corrected)
ceh_28_corrected <- normalize(ceh_28_corrected)
ceh_28_corrected
saveRDS(ceh_28_corrected, "./final_gene_model/4170-ST-1/SoupX2/020620_ceh_28_sce_new_SoupX_corrected.rds")

